// progressive-web-sdk/dist/store/checkout/shipping/selectors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getIsInitialized = exports.getCity = exports.getCountryId = exports.getRegionId = exports.getCompany = exports.getPostcode = exports.getTelephone = exports.getAddressLineTwo = exports.getAddressLineOne = exports.getStreet = exports.getShippingFullName = exports.getShippingLastName = exports.getShippingFirstName = exports.getSelectedShippingLabel = exports.getSelectedShippingRate = exports.getShippingAddressCustomContent = exports.getSelectedShippingMethod = exports.getInitialShippingAddress = exports.getSelectedShippingMethodValue = exports.getSelectedSavedAddressId = exports.getShippingAddress = undefined;

var _reselect = require('reselect');

var _immutable = require('immutable');

var _immutable2 = _interopRequireDefault(_immutable);

var _reselectImmutableHelpers = require('reselect-immutable-helpers');

var _selectors = require('../selectors');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var getShippingAddress = exports.getShippingAddress = (0, _reselectImmutableHelpers.createGetSelector)(_selectors.getCheckout, 'shippingAddress', _immutable2.default.Map());
var getSelectedSavedAddressId = exports.getSelectedSavedAddressId = (0, _reselectImmutableHelpers.createGetSelector)(_selectors.getCheckout, 'defaultShippingAddressId');
var getSelectedShippingMethodValue = exports.getSelectedShippingMethodValue = (0, _reselectImmutableHelpers.createGetSelector)(_selectors.getCheckout, 'selectedShippingMethodId', '');

var getInitialShippingAddress = exports.getInitialShippingAddress = (0, _reselect.createSelector)(_selectors.getCheckout, _selectors.getEmailAddress, getShippingAddress, getSelectedSavedAddressId, getSelectedShippingMethodValue, function (checkout, email, address, savedAddressId, shippingMethodId) {
    address = address.set('shippingMethodId', shippingMethodId).set('username', email);
    if (savedAddressId) {
        return address.set('savedAddress', '' + savedAddressId);
    }
    return address;
});

var getSelectedShippingMethod = exports.getSelectedShippingMethod = (0, _reselect.createSelector)(_selectors.getShippingMethods, getSelectedShippingMethodValue, function (shippingMethods, selectedMethodValue) {
    if (!shippingMethods.size) {
        return _immutable2.default.Map();
    }
    var selectedValue = shippingMethods.filter(function (method) {
        return method.get('id') === selectedMethodValue;
    });
    return selectedValue.size ? selectedValue.get(0) : shippingMethods.get(0);
});

var getShippingAddressCustomContent = exports.getShippingAddressCustomContent = (0, _reselectImmutableHelpers.createGetSelector)(getShippingAddress, 'custom', _immutable2.default.Map());

var getSelectedShippingRate = exports.getSelectedShippingRate = (0, _reselectImmutableHelpers.createGetSelector)(getSelectedShippingMethod, 'cost', '');

var getSelectedShippingLabel = exports.getSelectedShippingLabel = (0, _reselectImmutableHelpers.createGetSelector)(getSelectedShippingMethod, 'label', '');

var getShippingFirstName = exports.getShippingFirstName = (0, _reselectImmutableHelpers.createGetSelector)(getShippingAddress, 'firstname', '');

var getShippingLastName = exports.getShippingLastName = (0, _reselectImmutableHelpers.createGetSelector)(getShippingAddress, 'lastname', '');

var getShippingFullName = exports.getShippingFullName = (0, _reselect.createSelector)(getShippingFirstName, getShippingLastName, function (firstName, lastName) {
    return firstName + ' ' + lastName;
});

var getStreet = exports.getStreet = (0, _reselectImmutableHelpers.createGetSelector)(getShippingAddress, 'street', _immutable2.default.List());

var getAddressLineOne = exports.getAddressLineOne = (0, _reselectImmutableHelpers.createGetSelector)(getShippingAddress, 'addressLine1');

var getAddressLineTwo = exports.getAddressLineTwo = (0, _reselectImmutableHelpers.createGetSelector)(getShippingAddress, 'addressLine2');

var getTelephone = exports.getTelephone = (0, _reselectImmutableHelpers.createGetSelector)(getShippingAddress, 'telephone');

var getPostcode = exports.getPostcode = (0, _reselectImmutableHelpers.createGetSelector)(getShippingAddress, 'postcode');

var getCompany = exports.getCompany = (0, _reselectImmutableHelpers.createGetSelector)(getShippingAddress, 'company');

var getRegionId = exports.getRegionId = (0, _reselectImmutableHelpers.createGetSelector)(getShippingAddress, 'regionId');

var getCountryId = exports.getCountryId = (0, _reselectImmutableHelpers.createGetSelector)(getShippingAddress, 'countryId');

var getCity = exports.getCity = (0, _reselectImmutableHelpers.createGetSelector)(getShippingAddress, 'city');

var getIsInitialized = exports.getIsInitialized = (0, _reselectImmutableHelpers.createGetSelector)(getShippingAddress, 'isInitialized');
