// progressive-web-sdk/dist/components/skeleton-inline/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Related components:
 *
 * * [SkeletonBlock](#!/SkeletonBlock)
 * * [SkeletonText](#!/SkeletonText)
 *
 * An inline skeleton component respects text properties such as line-height
 * and font-size (which effectively becomes the height).
 */
var SkeletonInline = function SkeletonInline(props) {
    var classes = (0, _classnames2.default)('pw-skeleton-inline', props.className);

    var styles = (0, _extends3.default)({
        fontSize: props.size,
        width: props.width
    }, props.style);

    var attributes = {
        role: 'presentation',
        className: classes,
        style: styles

        // Add an href to the element's attributes if we're rendering an anchor
    };if (props.type === 'a') {
        attributes.href = '#';
    }

    return _react2.default.createElement(props.type, attributes);
}; /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

SkeletonInline.defaultProps = {
    size: 'inherit',
    type: 'span',
    width: '100%'
};

SkeletonInline.propTypes = {
    /**
    * Adds values to the `class` attribute of the root element.
    */
    className: _propTypes2.default.string,

    /**
    * Specifies the font-size (effectively the height) of the skeleton, in any valid dimension type.
    */
    size: _propTypes2.default.string,

    /**
    * Specifies any inline styles to be applied to the element.
    */
    style: _propTypes2.default.object,

    /**
    * Specifies the element type to be constructed.
    */
    type: _propTypes2.default.string,

    /**
    * Specifies the width of the skeleton's children's width, in any valid dimension type.
    */
    width: _propTypes2.default.string
};

exports.default = SkeletonInline;
