// progressive-web-sdk/dist/components/progress-steps/progress-steps-item.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _icon = require('../icon');

var _icon2 = _interopRequireDefault(_icon);

var _progressStepsStep = require('./partials/progress-steps-step');

var _progressStepsStep2 = _interopRequireDefault(_progressStepsStep);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Related component:
 *
 * * [ProgressSteps](#!/ProgressSteps)
 *
 * This component is used inside `ProgressSteps` component to display each step.
 *
 * @example ./DESIGN.md
 */

var ProgressStepsItem = function ProgressStepsItem(props) {
    var className = props.className,
        completed = props.completed,
        current = props.current,
        getStepMessage = props.getStepMessage,
        href = props.href,
        icon = props.icon,
        number = props.number,
        title = props.title,
        total = props.total;


    var classes = (0, _classnames2.default)('pw-progress-steps__item', {
        'pw--current': current,
        'pw--completed': completed
    }, className);

    var stepWidth = 100 / total;
    var styles = { width: stepWidth + '%' };

    return _react2.default.createElement(
        'li',
        { className: classes, style: styles },
        _react2.default.createElement(
            _progressStepsStep2.default,
            { className: 'pw-progress-steps__step', href: href },
            _react2.default.createElement(
                'div',
                { className: 'pw-progress-steps__label' },
                _react2.default.createElement(
                    'div',
                    { className: 'pw-progress-steps__number' },
                    _react2.default.createElement(
                        'div',
                        { className: 'pw-progress-steps__badge' },
                        icon ? _react2.default.createElement(_icon2.default, {
                            className: 'pw-progress-steps__icon',
                            name: icon
                        }) : '' + number
                    )
                ),
                _react2.default.createElement(
                    'div',
                    { className: 'pw-progress-steps__title' },
                    title,
                    _react2.default.createElement(
                        'span',
                        { className: 'u-visually-hidden' },
                        getStepMessage(number, total)
                    )
                )
            )
        )
    );
}; /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var defaultGetStepMessage = function defaultGetStepMessage(stepIndex, total) {
    return 'Step ' + stepIndex + ' of ' + total;
};

ProgressStepsItem.defaultProps = {
    completed: false,
    getStepMessage: defaultGetStepMessage
};

ProgressStepsItem.propTypes = {
    /**
     * The label for the step.
     */
    title: _propTypes2.default.string.isRequired,

    /**
     * Adds values to the `class` attribute of the root element.
     */
    className: _propTypes2.default.string,

    /**
     * PROVIDED INTERNALLY: Adds `pw--completed` styling class to completed step.
     */
    completed: _propTypes2.default.bool,

    /**
     * Indicates whether the step is the current one.
     */
    current: _propTypes2.default.bool,

    /**
     * This function should return a string (or node) that describes the meaning of each individual
     * "step" of the `ProgressSteps`. This function is run for each step, and is passed the current
     * step index and the total number of steps.
     */
    getStepMessage: _propTypes2.default.func,

    /**
     * If provided, the ProgressStepsItem will render as a link to it.
     */
    href: _propTypes2.default.string,

    /**
     * Icon to be displayed in the ProgressStepsItem.
     * If omitted, the step number will be displayed instead.
     * For more information about available icons, see the [Icon component](#!/Icon).
     */
    icon: _propTypes2.default.string,

    /**
     * PROVIDED INTERNALLY: The number of the step. Rendered if no icon is provided.
     */
    number: _propTypes2.default.number,

    /**
     * PROVIDED INTERNALLY: Total number of steps in the ProgressSteps.
     */
    total: _propTypes2.default.number
};

exports.default = ProgressStepsItem;
