// progressive-web-sdk/dist/components/progress-steps/progress-steps.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Related component:
 *
 * * [ProgressStepsItem](#!/ProgressStepsItem)
 *
 * `ProgressStep` component is commonly used in the checkout flow showing the user
 * which step they are currently on and an overview of the amount of steps in this workflow.
 *
 * @example ./DESIGN.md
 */
var ProgressSteps = function ProgressSteps(_ref) {
    var className = _ref.className,
        children = _ref.children;

    // count of steps
    var total = children.length;
    var completed = true;
    var classes = (0, _classnames2.default)('pw-progress-steps', className);

    return _react2.default.createElement(
        'ol',
        { className: classes },
        _react2.default.Children.map(children, function (child, index) {
            if (child.props.current) {
                completed = false;
            }

            var childProps = {
                total: total,
                number: index + 1,
                key: index,
                completed: completed
            };

            return _react2.default.cloneElement(child, childProps);
        })
    );
}; /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

ProgressSteps.propTypes = {
    /**
     * The set of steps in the ProgressSteps. To be provided as
     * ProgressStepsItem components.
     */
    children: _propTypes2.default.node.isRequired,

    /**
     * Adds values to the `class` attribute of the root element.
     */
    className: _propTypes2.default.string
};

exports.default = ProgressSteps;
