// progressive-web-sdk/dist/integration-manager/products/types.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.ProductUIData = exports.Products = undefined;

var _runtypes = require('runtypes');

var Runtypes = _interopRequireWildcard(_runtypes);

var _types = require('../types');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var VariationCategoryID = _types.Identifier;
var VariationCategoryName = _types.Identifier;

var Link = Runtypes.Record({
    href: _types.URL,
    text: _types.Text
}).And(Runtypes.Partial({
    title: _types.Text
}));

var Option = Runtypes.Record({
    value: _types.Identifier,
    label: _types.Text
});

var VariationCategory = Runtypes.Record({
    id: VariationCategoryID,
    name: VariationCategoryName,
    label: _types.Text,
    values: Runtypes.Array(Option)
});

var Variant = Runtypes.Record({
    id: _types.ProductID,
    values: Runtypes.Dictionary(VariationCategoryID, _types.Identifier)
});

var Product = Runtypes.Record({
    id: _types.ProductID,
    title: _types.Text,
    price: _types.Money,
    href: _types.URL,
    available: Runtypes.Boolean
}).And(Runtypes.Partial({
    pageMeta: _types.Meta,
    thumbnail: _types.Image,
    images: Runtypes.Array(_types.Image),
    description: _types.Text,
    variationCategories: Runtypes.Array(VariationCategory),
    variants: Runtypes.Array(Variant)
}));

var Products = exports.Products = Runtypes.Dictionary(Product, _types.ProductID);

var ProductUIData = exports.ProductUIData = Runtypes.Record({
    breadcrumbs: Runtypes.Array(Link),
    itemQuantity: Runtypes.Number
});
