// mobify-integration-manager/dist/connectors/hybris/categories/commands.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getCategories = exports.getCategory = undefined;

var _results = require('../../../integration-manager/api/categories/results');

var _config = require('../config');

var _utils = require('../utils');

var _utils2 = require('../../../utils/utils');

/**
 * @param {string} path - category path
 * @example "/cat/clothes/tshirts/270100"
 *
 * @param {string} catId - category id
 * @example "270100"
 *
 * @return {string} - path of parent category
 * @example "/cat/clothes/tshirts"
 */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2017 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var getParentCategoryPath = function getParentCategoryPath(path, catId) {
    if (!path || !catId) {
        return null;
    }

    var index = path.indexOf('/' + catId);
    var catPath = path.slice(0, index);

    if (catPath === '/' + _config.PATHS.CATEGORY) {
        return null;
    }

    return catPath;
};

var getCategory = exports.getCategory = function getCategory(id) {
    return function (dispatch) {
        var categoryInfoEndpoint = (0, _config.getCategoryEndPoint)(id);
        return (0, _utils.makeApiRequest)(categoryInfoEndpoint, { method: 'GET' }).then(function (response) {
            return response.json();
        }).then(function (responseJSON) {
            var id = responseJSON.id,
                name = responseJSON.name,
                url = responseJSON.url;

            var parentId = getParentCategoryPath(url, id);

            dispatch((0, _results.receiveCategoryInformation)(id, {
                id: id,
                title: name,
                href: url,
                parentId: parentId
            }));
        });
    };
};

var getCategories = exports.getCategories = _utils2.notImplementedCommand;
