// progressive-web-sdk/dist/store/app/selectors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getAvailableCurrencies = exports.getSelectedCurrency = exports.isStandaloneApp = exports.getCurrentPathKeyWithoutQuery = exports.getCurrentPathKey = exports.getCurrentUrl = exports.getApp = undefined;

var _immutable = require('immutable');

var _immutable2 = _interopRequireDefault(_immutable);

var _reselect = require('reselect');

var _reselectImmutableHelpers = require('reselect-immutable-helpers');

var _utils = require('../../utils/utils');

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var getApp = exports.getApp = function getApp(_ref) {
  var app = _ref.app;
  return app;
}; /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var getCurrentUrl = exports.getCurrentUrl = (0, _reselectImmutableHelpers.createGetSelector)(getApp, _constants.CURRENT_URL);
var getCurrentPathKey = exports.getCurrentPathKey = (0, _reselect.createSelector)(getCurrentUrl, _utils.urlToPathKey);
var getCurrentPathKeyWithoutQuery = exports.getCurrentPathKeyWithoutQuery = (0, _reselect.createSelector)(getCurrentPathKey, _utils.urlToBasicPathKey);
var isStandaloneApp = exports.isStandaloneApp = (0, _reselectImmutableHelpers.createGetSelector)(getApp, 'standaloneApp');
var getSelectedCurrency = exports.getSelectedCurrency = (0, _reselectImmutableHelpers.createGetSelector)(getApp, 'selectedCurrency', {});
var getAvailableCurrencies = exports.getAvailableCurrencies = (0, _reselectImmutableHelpers.createGetSelector)(getApp, 'availableCurrencies', _immutable2.default.List());
