// progressive-web-sdk/dist/analytics/connectors/google-tag-manager/connector.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _get2 = require('babel-runtime/helpers/get');

var _get3 = _interopRequireDefault(_get2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _connector = require('../connector');

var _connector2 = _interopRequireDefault(_connector);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var GoogleTagManager = function (_Connector) {
    (0, _inherits3.default)(GoogleTagManager, _Connector);

    function GoogleTagManager() {
        var displayName = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Google Tag Manager';
        var containerId = arguments[1];
        var options = arguments[2];
        (0, _classCallCheck3.default)(this, GoogleTagManager);

        var _this = (0, _possibleConstructorReturn3.default)(this, (GoogleTagManager.__proto__ || (0, _getPrototypeOf2.default)(GoogleTagManager)).call(this, displayName, options));

        if (!containerId) {
            throw new Error('containerId cannot be undefined');
        }

        window.dataLayer = [{
            'gtm.start': new Date().getTime(),
            event: 'gtm.js'
        }];

        // Load Google Analytics
        _connector2.default.loadScript('https://www.googletagmanager.com/gtm.js?id=' + containerId, _this);
        return _this;
    }

    (0, _createClass3.default)(GoogleTagManager, [{
        key: 'debug',
        value: function debug(type, payload) {
            var _console;

            (_console = console).groupCollapsed.apply(_console, (0, _toConsumableArray3.default)(_connector2.default.debugHeading('' + this.displayName, type)));
            console.log('DataLayer ' + payload);
            console.groupEnd();
        }
    }, {
        key: 'pageviewEvent',
        value: function pageviewEvent(payload) {
            return (0, _extends3.default)({
                event: 'Pageview'
            }, payload);
        }
    }, {
        key: 'purchaseEvent',
        value: function purchaseEvent(payload) {
            return {
                event: 'Ecommerce',
                ec: {
                    purchase: {
                        actionField: payload.transaction,
                        products: payload.products
                    }
                }
            };
        }
    }, {
        key: 'send',
        value: function send(type, payload) {
            window.dataLayer.push(payload);

            (0, _get3.default)(GoogleTagManager.prototype.__proto__ || (0, _getPrototypeOf2.default)(GoogleTagManager.prototype), 'send', this).call(this, type, payload);
        }
    }]);
    return GoogleTagManager;
}(_connector2.default); /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

exports.default = GoogleTagManager;
