// progressive-web-sdk/dist/components/ledger/ledger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _ledgerRow = require('./ledger-row');

var _ledgerRow2 = _interopRequireDefault(_ledgerRow);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Related component:
 *
 * * [LedgerRow](#!/LedgerRow)
 *
 * A table/ledger component used to display order summaries.
 *
 * @example ./DESIGN.md
 */

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var Ledger = function Ledger(_ref) {
    var children = _ref.children,
        className = _ref.className,
        rows = _ref.rows;


    var classes = (0, _classnames2.default)('pw-ledger', className);

    if (children) {
        return _react2.default.createElement(
            'table',
            { className: classes },
            _react2.default.createElement(
                'tbody',
                null,
                children
            )
        );
    }

    return _react2.default.createElement(
        'table',
        { className: classes },
        _react2.default.createElement(
            'tbody',
            null,
            rows.map(function (row, idx) {
                return _react2.default.createElement(_ledgerRow2.default, (0, _extends3.default)({}, row, { key: idx }));
            })
        )
    );
};

var ledgerRowPropType = function ledgerRowPropType(props, propName, componentName) {
    var error = void 0;
    var prop = props[propName];

    _react2.default.Children.forEach(prop, function (child) {
        if (!!child && child.type.displayName !== 'LedgerRow') {
            error = new Error(componentName + ' only accepts children of type "LedgerRow"');
        }
    });
    return error;
};

Ledger.propTypes = {
    /**
    * Content for the ledger, must be of type LedgerRow.
    */
    children: ledgerRowPropType,

    /**
    * Additional classes to add to the ledger container.
    */
    className: _propTypes2.default.string,

    /**
    * A collection of objects defining the content in each entry in the Ledger.
    * (See Details of rows prop below for more info)
    */
    rows: _propTypes2.default.arrayOf(_propTypes2.default.shape({
        isTotal: _propTypes2.default.bool,
        label: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node]),
        value: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node])
    }))
};

exports.default = Ledger;
