// progressive-web-sdk/dist/components/ledger/ledger-row.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Related component:
 *
 * * [Ledger](#!/Ledger)
 *
 * This is used inside ledger component to display each row of the table.
 *
 * @example ./DESIGN.md
 */

var LedgerRow = function LedgerRow(_ref) {
    var className = _ref.className,
        isTotal = _ref.isTotal,
        label = _ref.label,
        labelAction = _ref.labelAction,
        labelDescription = _ref.labelDescription,
        value = _ref.value,
        valueAction = _ref.valueAction;

    var classes = (0, _classnames2.default)('pw-ledger__row', {
        'pw--total': isTotal
    }, className);

    return _react2.default.createElement(
        'tr',
        { className: classes },
        _react2.default.createElement(
            'td',
            { className: 'pw-ledger__item' },
            label,
            (labelDescription || labelAction) && _react2.default.createElement(
                'div',
                { className: 'pw-ledger__details' },
                labelDescription,
                _react2.default.createElement(
                    'div',
                    { className: 'pw-ledger__action' },
                    labelAction
                )
            )
        ),
        _react2.default.createElement(
            'td',
            { className: 'pw-ledger__value' },
            value,
            valueAction && _react2.default.createElement(
                'div',
                { className: 'pw-ledger__details' },
                _react2.default.createElement(
                    'div',
                    { className: 'pw-ledger__action' },
                    valueAction
                )
            )
        )
    );
}; /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

LedgerRow.displayName = 'LedgerRow';

LedgerRow.propTypes = {
    /**
    * Additional classes to add to the ledgerRow container.
    */
    className: _propTypes2.default.string,

    /**
    * Indicates whether this ledger entry is the final total row.
    */
    isTotal: _propTypes2.default.bool,

    /**
    * The label for the ledger entry.
    */
    label: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node]),

    /**
    * Additional content that appears within the `pw-ledger__item` container.
    */
    labelAction: _propTypes2.default.node,

    /**
    * Label description for ledger entry.
    */
    labelDescription: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node]),

    /**
    * The value for the ledger entry.
    */
    value: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node]),

    /**
    * Additional content that appears within the `pw-ledger__value` container.
    */
    valueAction: _propTypes2.default.node
};

exports.default = LedgerRow;
