// progressive-web-sdk/src/utils/messaging-state.js
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

import {DEFAULT_CHANNEL, VISIT_COUNTDOWNS, PERMA_DURATION, PAGE_COUNT, ACTIVE_VISIT_DURATION} from '../store/push-messaging/constants'
import Logger from './logger'
import PushMessagingStore from '../store/push-messaging/push-messaging-store'

let storage
let logger

/**
 * Creates local instances of PushMessaginStore and Logger, for use in other
 * methods. Call this first before any other methods.
 */
export const initStorage = () => {
    storage = storage || new PushMessagingStore('pw')
    logger = logger || new Logger('[Messaging UI]')

    return storage
}
/**
 * Sets the provided countdowns object in local storage.
 *
 * @param {object} countdowns - The visit countdowns object to set in storage
 * @returns {object} the provided `countdowns` argument
 */
export const setVisitCountdownsInStorage = (countdowns) => {
    storage.set(VISIT_COUNTDOWNS, countdowns, PERMA_DURATION)

    return countdowns
}

/**
 * Gets the visit countdowns object from local storage
 *
 * @returns {object} current visit countdowns
 */
export const getVisitCountdowns = () => {
    return storage.get(VISIT_COUNTDOWNS) || {}
}

/**
 * Adds the optionally provided channel name, or "default" to the visit countdowns
 * object, with the provided `visitsToWait` argument, which is then set in local
 * storage.
 *
 * @param {number} visitsToWait - The number of visits to wait before showing an ask again
 * @param {string} [channelName] - Optional channel name (default is 'broadcast')
 * @returns {object} current visit countdowns
 */
export const startVisitCountdown = (visitsToWait, channelName = DEFAULT_CHANNEL) => {
    const countdowns = getVisitCountdowns()
    countdowns[channelName] = visitsToWait

    return setVisitCountdownsInStorage(countdowns)
}

/**
 * Retrieves visit countdowns from local storage, decreases each key's value by 1,
 * then sets the value in local storage again.
 *
 * @returns {object} - current visit countdowns
 */
export const decreaseVisitCountdowns = () => {
    const countdowns = getVisitCountdowns()

    for (const counter in countdowns) {
        /* istanbul ignore else */
        if (countdowns.hasOwnProperty(counter)) {
            countdowns[counter] = Math.max(0, countdowns[counter] - 1)
        }
    }

    return setVisitCountdownsInStorage(countdowns)
}

/**
 * Gets the current page count from local storage.
 *
 * @returns {number} The current page count, or 0 if not found in storage
 */
export const getPageCount = () => {
    return storage.get(PAGE_COUNT) || 0
}

/**
 * Sets the page count in storage using the provided pageCount argument
 *
 * @param {number} pageCount - the page count to set in storage
 * @returns {number} the provided `pageCount` argument
 */
export const setPageCount = (pageCount) => {
    return storage.set(PAGE_COUNT, pageCount, ACTIVE_VISIT_DURATION)
}

/**
 * Updates and retrieves the current page count and visit countdowns from local
 * storage, decreasing visit countdowns if page count was not found.
 *
 * @returns {object}
 *   - pageCount: the current page count
 *   - visitCountdowns: the current visit countdowns
 */
export const updateState = () => {
    let pageCount = getPageCount()
    let countdowns

    // If pageCount is 0, it means it wasn't found in storage - this indicates
    // it expired and we should decrement visit countdowns
    if (pageCount === 0) {
        logger.log('Visit elapsed, decreasing visit countdowns.')
        countdowns = decreaseVisitCountdowns()
    } else {
        countdowns = getVisitCountdowns()
    }

    setPageCount(++pageCount)

    return {
        pageCount,
        visitCountdowns: countdowns
    }
}
