// progressive-web-sdk/src/utils/preview-utils.js
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

import {loadScript} from './utils'

const PREVIEW_URL = 'https://preview.mobify.com/v7/'

export const getCookie = (name) => {
    // Internet Explorer treats empty cookies differently, it does
    // not include the '=', so our regex has to be extra complicated.
    const re = new RegExp(`(?:^|; )${name}(?:(?:=([^;]*))|(?:; |$))`)
    const match = document.cookie.match(re)
    return (match && match[1]) || ''
}

export const isPreview = () => {
    return (getCookie('mobify-path') === 'true' ||
        /mobify-path=true/.test(window.location.hash))
}

export const isV8Tag = () => {
    if (window.Mobify.tagVersion && window.Mobify.tagVersion[0] === 8) {
        return true
    }
    return false
}

export const SHOULD_PREVIEW_QUERY = `script[src="${PREVIEW_URL}"]`
export const shouldPreview = () => {
    // If the tag is not a V8 tag, then the tag itself will deal with previewing.
    if (!isV8Tag()) {
        return false
    }

    const previewHasRun = document.querySelectorAll(SHOULD_PREVIEW_QUERY).length > 0
    if (previewHasRun) {
        return false
    }

    return isPreview()
}

export const loadPreview = /* istanbul ignore next */ () => {
    /* istanbul ignore next */
    loadScript({
        src: PREVIEW_URL,
        docwrite: window.Mobify.shouldLoadPWA,
        isAsync: true
    })
}
