// progressive-web-sdk/dist/store/user/orders/selectors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getOrderShippingAddress = exports.getOrderShippingMethod = exports.getOrderPaymentMethods = exports.getOrderBillingAddress = exports.getOrderSubtotal = exports.getOrderShippingTotal = exports.getOrderTax = exports.getOrderTotal = exports.getOrderItems = exports.getOrderStatus = exports.getOrderDate = exports.getOrderPageMeta = exports.getCurrentOrder = exports.getOrders = exports.getCurrentOrderNumber = undefined;

var _immutable = require('immutable');

var _immutable2 = _interopRequireDefault(_immutable);

var _reselect = require('reselect');

var _reselectImmutableHelpers = require('reselect-immutable-helpers');

var _selectors = require('../selectors');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
// /* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
// /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
var getCurrentOrderNumber = exports.getCurrentOrderNumber = (0, _reselectImmutableHelpers.createGetSelector)(_selectors.getUser, 'currentOrderNumber', '');
var getOrders = exports.getOrders = (0, _reselectImmutableHelpers.createGetSelector)(_selectors.getUser, 'orders', _immutable2.default.Map());

var getCurrentOrder = exports.getCurrentOrder = (0, _reselect.createSelector)(getOrders, getCurrentOrderNumber, function (orders, currentOrdeNumber) {
    return orders.get(currentOrdeNumber) || _immutable2.default.Map();
});

var getOrderPageMeta = exports.getOrderPageMeta = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentOrder, 'pageMeta', _immutable2.default.Map());
var getOrderDate = exports.getOrderDate = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentOrder, 'date');
var getOrderStatus = exports.getOrderStatus = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentOrder, 'status');
var getOrderItems = exports.getOrderItems = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentOrder, 'items');
var getOrderTotal = exports.getOrderTotal = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentOrder, 'total');
var getOrderTax = exports.getOrderTax = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentOrder, 'tax');
var getOrderShippingTotal = exports.getOrderShippingTotal = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentOrder, 'shippingTotal');
var getOrderSubtotal = exports.getOrderSubtotal = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentOrder, 'subtotal');
var getOrderBillingAddress = exports.getOrderBillingAddress = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentOrder, 'billingAddress');
var getOrderPaymentMethods = exports.getOrderPaymentMethods = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentOrder, 'paymentMethods', _immutable2.default.List());
var getOrderShippingMethod = exports.getOrderShippingMethod = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentOrder, 'shippingMethod');
var getOrderShippingAddress = exports.getOrderShippingAddress = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentOrder, 'shippingAddress');
