// progressive-web-sdk/dist/store/modals/actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.preRenderModal = exports.persistModal = exports.closeAllModals = exports.openPersistentModal = exports.closeModal = exports.openModal = undefined;

var _actionCreation = require('../../utils/action-creation');

var _dataObjects = require('../../analytics/data-objects/');

var _actions = require('../../analytics/actions');

var openModal = exports.openModal = (0, _actionCreation.createActionWithAnalytics)('Open modal', ['modalName'], _dataObjects.EVENT_ACTION.uiInteraction, function (name, analyticsName) {
    return (0, _actions.createModalAnalyticsMeta)(_dataObjects.UI_ACTION.open, name, analyticsName);
}); /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var closeModal = exports.closeModal = (0, _actionCreation.createActionWithAnalytics)('Close modal', ['modalName'], _dataObjects.EVENT_ACTION.uiInteraction, function (name, analyticsName) {
    return (0, _actions.createModalAnalyticsMeta)(_dataObjects.UI_ACTION.close, name, analyticsName);
});

// A persistent modal will not be closed after route change
var openPersistentModal = exports.openPersistentModal = (0, _actionCreation.createActionWithAnalytics)('Open persistent modal', ['modalName'], _dataObjects.EVENT_ACTION.uiInteraction, function (name, analyticsName) {
    return (0, _actions.createModalAnalyticsMeta)(_dataObjects.UI_ACTION.open, name, analyticsName);
});

var closeAllModals = exports.closeAllModals = (0, _actionCreation.createActionWithAnalytics)('Close all modals', [], _dataObjects.EVENT_ACTION.uiInteraction, function () {
    return (0, _actions.createModalAnalyticsMeta)(_dataObjects.UI_ACTION.close, _dataObjects.UI_NAME.all);
});

var persistModal = exports.persistModal = (0, _actionCreation.createAction)('Persist modal', ['modalName']);
var preRenderModal = exports.preRenderModal = (0, _actionCreation.createAction)('Pre-render modal', ['modalName']);
