// progressive-web-sdk/dist/store/modals/selectors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.isModalOpen = exports.getPreRenderingModals = exports.getClosedModals = exports.getOpenModals = exports.getKnownModals = exports.getModals = undefined;

var _reselect = require('reselect');

// import {createGetSelector} from 'reselect-immutable-helpers'

var getModals = exports.getModals = function getModals(_ref) {
    var modals = _ref.modals;
    return modals;
};

// @TODO: Deprecate getKnownModals
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var getKnownModals = exports.getKnownModals = function getKnownModals(_ref2) {
    var modals = _ref2.modals;
    return modals;
};

var getOpenModals = exports.getOpenModals = (0, _reselect.createSelector)(getModals, function (modals) {
    return modals.filter(function (modal) {
        return modal.get('open');
    });
});

var getClosedModals = exports.getClosedModals = (0, _reselect.createSelector)(getModals, function (modals) {
    return modals.filterNot(function (modal) {
        return modal.get('open');
    });
});

var getPreRenderingModals = exports.getPreRenderingModals = (0, _reselect.createSelector)(getClosedModals, function (modals) {
    return modals.filter(function (modal) {
        return modal.get('prerender');
    });
});

var isModalOpen = exports.isModalOpen = function isModalOpen(modalName) {
    return (0, _reselect.createSelector)(getOpenModals, function (modals) {
        return modals.getIn([modalName, 'open']) || false;
    });
};
