// mobify-integration-manager/dist/constants/form/constants.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2017 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var ADD_TO_CART_FORM_NAME = exports.ADD_TO_CART_FORM_NAME = 'product-add-to-cart';
var CART_PROMO_FORM_NAME = exports.CART_PROMO_FORM_NAME = 'cartPromoForm';
var CONFIRMATION_FORM_NAME = exports.CONFIRMATION_FORM_NAME = 'confirmationForm';
var ESTIMATE_FORM_NAME = exports.ESTIMATE_FORM_NAME = 'estimateShipping';
var NEWSLETTER_FORM_NAME = exports.NEWSLETTER_FORM_NAME = 'newsletterForm';
var PAYMENT_FORM_NAME = exports.PAYMENT_FORM_NAME = 'paymentForm';
var REGISTER_FORM_NAME = exports.REGISTER_FORM_NAME = 'register-form';
var SHIPPING_FORM_NAME = exports.SHIPPING_FORM_NAME = 'shippingForm';
var ADDRESS_FORM_NAME = exports.ADDRESS_FORM_NAME = 'addressForm';
var SIGN_IN_FORM_NAME = exports.SIGN_IN_FORM_NAME = 'signin-form';
