// progressive-web-sdk/dist/components/review-summary/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _image = require('../image');

var _image2 = _interopRequireDefault(_image);

var _rating = require('../rating');

var _rating2 = _interopRequireDefault(_rating);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * `ReviewSummary` component is used to show overall rating and summary.
 * It uses rating and image components.
 *
 * @example ./DESIGN.md
 */

var ReviewSummary = function ReviewSummary(_ref) {
    var className = _ref.className,
        imageProps = _ref.imageProps,
        ratingProps = _ref.ratingProps,
        summary = _ref.summary,
        text = _ref.text;

    var classes = (0, _classnames2.default)('pw-review-summary', className);
    var ratingStarsClasses = (0, _classnames2.default)('pw-review-summary__rating-stars', ratingProps.className);

    // alt is in imageProps
    /* eslint-disable jsx-a11y/img-has-alt */
    return _react2.default.createElement(
        'div',
        { className: classes },
        _react2.default.createElement(
            'div',
            { className: 'pw-review-summary__image' },
            _react2.default.createElement(_image2.default, imageProps)
        ),
        _react2.default.createElement(
            'div',
            { className: 'pw-review-summary__rating' },
            _react2.default.createElement(
                'div',
                { className: 'pw-review-summary__rating-inner' },
                _react2.default.createElement(_rating2.default, (0, _extends3.default)({}, ratingProps, { className: ratingStarsClasses })),
                _react2.default.createElement(
                    'div',
                    { className: 'pw-review-summary__rating-info' },
                    summary
                )
            ),
            _react2.default.createElement(
                'div',
                { className: 'pw-review-summary__recommended' },
                text
            )
        )
    );
}; /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

ReviewSummary.defaultProps = {
    ratingProps: {
        className: 'pw--solid'
    }
};

ReviewSummary.propTypes = {
    /**
     * Adds values to the `class` attribute of the root element.
     */
    className: _propTypes2.default.string,

    /**
     * The properties for Image.
     */
    imageProps: _propTypes2.default.object,

    /**
     * The properties for Rating.
     */
    ratingProps: _propTypes2.default.object,

    /**
     * Summary of the review that displays the number of given stars over total stars.
     */
    summary: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node]),

    /**
     * Content of the review summary.
     */
    text: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node])
};

exports.default = ReviewSummary;
