// mobify-integration-manager/dist/integration-manager/api/account/results.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.receiveSavedAddresses = exports.receiveWishlistCustomContent = exports.receiveUpdatedWishlistItem = exports.removeWishlistItem = exports.receiveWishlistData = exports.receiveCurrentOrderNumber = exports.receiveAccountOrderListData = exports.receiveAccountInfoData = undefined;

var _actionCreation = require('progressive-web-sdk/dist/utils/action-creation');

var receiveAccountInfoData = exports.receiveAccountInfoData = (0, _actionCreation.createAction)('Receive Account Info Data'); /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2017 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var receiveAccountOrderListData = exports.receiveAccountOrderListData = (0, _actionCreation.createAction)('Receive Account Order List Data', ['orders']);

var receiveCurrentOrderNumber = exports.receiveCurrentOrderNumber = (0, _actionCreation.createAction)('Receive Current Order Id', ['currentOrderNumber']);

var receiveWishlistData = exports.receiveWishlistData = (0, _actionCreation.createAction)('Receive wishlist data', ['wishlist']);
var removeWishlistItem = exports.removeWishlistItem = (0, _actionCreation.createAction)('Remove Wishlist Item', ['removeItemId']);
var receiveUpdatedWishlistItem = exports.receiveUpdatedWishlistItem = (0, _actionCreation.createAction)('Receives a updated Wishlist Item data');
var receiveWishlistCustomContent = exports.receiveWishlistCustomContent = (0, _actionCreation.createAction)('Receive Wishlist custom content');

var receiveSavedAddresses = exports.receiveSavedAddresses = (0, _actionCreation.createAction)('Receive Saved Addresses', ['savedAddresses']);
