// progressive-web-sdk/dist/components/swatch/swatch.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _swatchItem = require('./swatch-item');

var _swatchItem2 = _interopRequireDefault(_swatchItem);

var _utils = require('../../utils/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Related component:
 *
 * * [SwatchItem](#!/SwatchItem)
 *
 * `Swatch` allows the user to select from a small list of options.
 *
 * @example ./DESIGN.md
 */

var Swatch = function Swatch(props) {
    var label = props.label,
        children = props.children,
        value = props.value,
        onChange = props.onChange,
        className = props.className;


    var classes = (0, _classnames2.default)('pw-swatch', className);
    var labelId = label.replace(/\s/g, '-');

    return _react2.default.createElement(
        'div',
        {
            className: classes,
            role: 'radiogroup',
            'aria-labelledby': 'swatch-label-' + labelId
        },
        _react2.default.createElement(
            'div',
            { className: 'pw-swatch__label', id: 'swatch-label-' + labelId },
            label
        ),
        _react2.default.createElement(
            'div',
            { className: 'pw-swatch__items' },
            _react2.default.Children.map(children, function (child) {
                if (child && child.type && child.type.name === _swatchItem2.default.name) {
                    var childValue = child.props.value;

                    return _react2.default.cloneElement(child, {
                        selected: childValue === value,
                        key: childValue,
                        onClick: onChange
                    });
                } else {
                    return child;
                }
            })
        )
    );
}; /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

Swatch.defaultProps = {
    label: '',
    onChange: _utils.noop
};

Swatch.propTypes = {
    /**
     * The options to choose between (Should be SwatchItems).
     */
    children: _propTypes2.default.node,

    /**
     * Adds values to the `class` attribute of the root element.
     */
    className: _propTypes2.default.string,

    /**
     * The label for the Swatch.
     */
    label: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node]),

    /**
     * The selected swatch value.
     */
    value: _propTypes2.default.string,

    /**
     * This function is called whenever the user selects a new value.
     * It is passed the new value.
     */
    onChange: _propTypes2.default.func
};

exports.default = Swatch;
