// progressive-web-sdk/src/components/bazaarvoice-wrapper/index.jsx
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

import React from 'react'
import PropTypes from 'prop-types'
import {getDisplayName} from '../../utils/component-utils'
import AppError from '../../analytics/error'

/**
 * This Higher Order Component can be used to construct
 * components that use Bazaarvoice.
 * It loads the Bazaarvoice API script once
 * and renders its wrapped component after the script has been loaded.
 * This means that the wrapped components can use componentDidMount
 * to run any initialization code that relies on the api script.
 */

const scriptClass = 'js-bvapi'
const loadingClass = 'pw--loading'
const loadedClass = 'pw--loaded'

const bazaarvoiceWrapper = (WrappedComponent) => {
    class BazaarvoiceWrapper extends React.Component {
        constructor(props) {
            super(props)

            this.state = {
                apiLoaded: false
            }

            this.WrappedComponent = WrappedComponent
        }

        componentDidMount() {
            const {
                apiSrc,
                configurationOptions,
                onError
            } = this.props

            const existingScript = document.querySelector(`.${scriptClass}`)

            if (!existingScript) {
                const script = document.createElement('script')
                script.src = apiSrc
                script.className = `${scriptClass} ${loadingClass}`
                script.onload = () => {
                    this.setState({
                        apiLoaded: true
                    })

                    window.$BV.configure('global', configurationOptions)
                    script.classList.remove(loadingClass)
                    script.classList.add(loadedClass)
                }
                script.onerror = () => {
                    onError()

                    try {
                        // This AppError automatically sends analytics about this error to our Engagement Engine
                        // However, we want to immediately catch it because we don't want to break the PWA
                        // (or make anyone looking at the console think we broke the PWA)
                        // just because a third party plugin failed to load
                        throw new AppError('Bazarrvoice Error')
                    } catch (e) {
                        console.error(e)
                    }
                }

                document.body.appendChild(script)
            } else if (new RegExp(loadedClass).test(existingScript.className)) {
                // Another instance has already loaded this script
                this.setState({
                    apiLoaded: true
                })
            }
        }

        render() {
            return this.state.apiLoaded ? <WrappedComponent {...this.props} /> : false
        }
    }
    BazaarvoiceWrapper.WrappedComponent = WrappedComponent
    BazaarvoiceWrapper.displayName = getDisplayName(WrappedComponent)
    BazaarvoiceWrapper.defaultProps = {
        configurationOptions: {},
        onError: () => {}
    }
    BazaarvoiceWrapper.propTypes = {
        apiSrc: PropTypes.string,
        configurationOptions: PropTypes.object,
        onError: PropTypes.func
    }

    return BazaarvoiceWrapper
}

export default bazaarvoiceWrapper
