// progressive-web-sdk/dist/components/review/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _rating = require('../rating');

var _rating2 = _interopRequireDefault(_rating);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * `Review` component is used to show title, author's info and descripton.
 * It also uses rating component to display user's satisfaction.
 *
 * @example ./DESIGN.md
 */

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var Review = function Review(_ref) {
    var text = _ref.text,
        className = _ref.className,
        ratingProps = _ref.ratingProps,
        info = _ref.info,
        title = _ref.title;

    var classes = (0, _classnames2.default)('pw-review', className);
    var ratingClasses = (0, _classnames2.default)('pw-review__rating', ratingProps.className);

    return _react2.default.createElement(
        'article',
        { className: classes },
        _react2.default.createElement(
            'header',
            { className: 'pw-review__header' },
            _react2.default.createElement(
                'h1',
                { className: 'pw-review__heading' },
                title
            ),
            _react2.default.createElement(_rating2.default, (0, _extends3.default)({}, ratingProps, { className: ratingClasses }))
        ),
        info && _react2.default.createElement(
            'div',
            { className: 'pw-review__author' },
            info.map(function (info, idx) {
                return _react2.default.createElement(
                    'p',
                    { className: 'pw-review__author-info', key: idx },
                    info
                );
            })
        ),
        _react2.default.createElement(
            'p',
            { className: 'pw-review__content' },
            text
        )
    );
};

Review.defaultProps = {
    ratingProps: {
        className: 'pw--solid'
    }
};

Review.propTypes = {
    /**
     * Adds values to the `class` attribute of the root element.
     */
    className: _propTypes2.default.string,

    /**
     * An array of information related to the author.
     */
    info: _propTypes2.default.array,

    /**
     * The properties for Rating.
     */
    ratingProps: _propTypes2.default.object,

    /**
     * Content of the review.
     */
    text: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node]),

    /**
     * Title of the review.
     */
    title: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node])
};

exports.default = Review;
