// progressive-web-sdk/dist/components/carousel/partials/carousel-button.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _button = require('../../button');

var _button2 = _interopRequireDefault(_button);

var _icon = require('../../icon');

var _icon2 = _interopRequireDefault(_icon);

var _dataObjects = require('../../../analytics/data-objects/');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CarouselButton = function CarouselButton(_ref) {
    var className = _ref.className,
        buttonClass = _ref.buttonClass,
        onClick = _ref.onClick,
        disabled = _ref.disabled,
        icon = _ref.icon,
        iconSize = _ref.iconSize,
        title = _ref.title,
        analyticsContent = _ref.analyticsContent;

    return _react2.default.createElement(
        'div',
        { className: className },
        _react2.default.createElement(
            _button2.default,
            {
                className: buttonClass,
                onClick: onClick,
                disabled: disabled,
                'data-analytics-name': _dataObjects.UI_NAME.carousel,
                'data-analytics-content': analyticsContent
            },
            _react2.default.createElement(_icon2.default, { name: icon, size: iconSize, title: title })
        )
    );
}; /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

CarouselButton.propTypes = {
    analyticsContent: _propTypes2.default.string,
    buttonClass: _propTypes2.default.string,
    className: _propTypes2.default.string,
    disabled: _propTypes2.default.bool,
    icon: _propTypes2.default.string,
    iconSize: _propTypes2.default.string,
    title: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node]),
    onClick: _propTypes2.default.func
};

exports.default = CarouselButton;
