// mobify-integration-manager/dist/connectors/hybris/checkout/parsers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.parseMobifyToHybrisPaymentDetails = undefined;

var _cardUtils = require('progressive-web-sdk/dist/card-utils');

var _config = require('../config');

var _utils = require('../../../utils/utils');

var parseMobifyToHybrisPaymentDetails = exports.parseMobifyToHybrisPaymentDetails = function parseMobifyToHybrisPaymentDetails() {
    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
        addressLine1 = _ref.addressLine1,
        _ref$addressLine = _ref.addressLine2,
        addressLine2 = _ref$addressLine === undefined ? '' : _ref$addressLine,
        ccexpiry = _ref.ccexpiry,
        ccname = _ref.ccname,
        ccnumber = _ref.ccnumber,
        city = _ref.city,
        countryId = _ref.countryId,
        name = _ref.name,
        postcode = _ref.postcode,
        regionId = _ref.regionId,
        telephone = _ref.telephone,
        titleCode = _ref.titleCode;

    var cardType = (0, _cardUtils.getCardData)(ccnumber).cardType;

    var _splitFullName = (0, _utils.splitFullName)(name),
        firstname = _splitFullName.firstname,
        lastname = _splitFullName.lastname;

    var address = {
        accountHolderName: ccname,
        'billingAddress.titleCode': titleCode,
        'billingAddress.firstName': firstname,
        'billingAddress.lastName': lastname,
        'billingAddress.line1': addressLine1,
        'billingAddress.line2': addressLine2,
        'billingAddress.postalCode': postcode,
        'billingAddress.town': city,
        'billingAddress.country.isocode': countryId,
        'billingAddress,region.isocode': regionId,
        'billingAddress.phone': telephone,
        cardNumber: ccnumber,
        cardType: _config.CREDIT_CARD_TYPES[cardType],
        expiryMonth: ccexpiry.substring(0, 2),
        expiryYear: ccexpiry.substring(2, 4)
    };
    return address;
}; /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2017 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
