// progressive-web-sdk/dist/store/push-messaging/selectors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getLocale = exports.isSystemAskShown = exports.getChannels = exports.getVisitCountdowns = exports.getPageCount = exports.getStatus = exports.canSubscribe = exports.isSubscribed = exports.isSupported = exports.isMessagingReady = exports.getPushMessaging = undefined;

var _reselectImmutableHelpers = require('reselect-immutable-helpers');

var _constants = require('./constants');

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var getPushMessaging = exports.getPushMessaging = function getPushMessaging(_ref) {
  var pushMessaging = _ref.pushMessaging;
  return pushMessaging;
};

var isMessagingReady = exports.isMessagingReady = (0, _reselectImmutableHelpers.createGetSelector)(getPushMessaging, 'isReady'); // Deprecated
var isSupported = exports.isSupported = (0, _reselectImmutableHelpers.createGetSelector)(getPushMessaging, 'isSupported', false);
var isSubscribed = exports.isSubscribed = (0, _reselectImmutableHelpers.createGetSelector)(getPushMessaging, 'subscribed');
var canSubscribe = exports.canSubscribe = (0, _reselectImmutableHelpers.createGetSelector)(getPushMessaging, 'canSubscribe');
var getStatus = exports.getStatus = (0, _reselectImmutableHelpers.createGetSelector)(getPushMessaging, 'status');

var getPageCount = exports.getPageCount = (0, _reselectImmutableHelpers.createGetSelector)(getPushMessaging, _constants.PAGE_COUNT);
var getVisitCountdowns = exports.getVisitCountdowns = (0, _reselectImmutableHelpers.createGetSelector)(getPushMessaging, _constants.VISIT_COUNTDOWNS);
var getChannels = exports.getChannels = (0, _reselectImmutableHelpers.createGetSelector)(getPushMessaging, 'channels');
var isSystemAskShown = exports.isSystemAskShown = (0, _reselectImmutableHelpers.createGetSelector)(getPushMessaging, 'systemAskShown');

var getLocale = exports.getLocale = (0, _reselectImmutableHelpers.createGetSelector)(getPushMessaging, 'locale');
