// progressive-web-sdk/src/components/image/index.jsx
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import {noop} from '../../utils/utils'

import Ratio from '../ratio'
import SkeletonBlock from '../skeleton-block'
import {PERFORMANCE_METRICS} from '../../analytics/data-objects/'
import {countAsset, trackPerformance} from '../../analytics/actions'

/**
 * An image component with placeholder functionality baked-in.
 * You can review the documentation for the [placeholder component here](#!/SkeletonBlock).
 * Note: The Image must have an explicit height for placeholder to appear.
 */
class Image extends React.PureComponent {
    constructor() {
        super()

        this.state = {
            loaded: false,
            transitioningImageProps: null
        }

        this.imageLoaded = this.imageLoaded.bind(this)
    }

    componentWillReceiveProps(nextProps) {
        if (this.props.src !== nextProps.src) {
            const {alt, height, src, width} = this.props
            const tagClasses = classNames('pw-image__tag', 'pw--is-transitioning')
            const imageProps = {
                src,
                className: tagClasses,
                alt,
                height,
                width
            }
            this.setState({loaded: false, transitioningImageProps: imageProps})
        }
    }

    componentWillMount() {
        if (this.props.src) {
            countAsset()
        }
    }

    componentWillUnmount() {
        // Clear the loading timeout so we don't call setState on an unmounted component
        if (this.timeout) {
            window.clearTimeout(this.timeout)
        }
    }

    imageLoaded() {
        trackPerformance(PERFORMANCE_METRICS.fullPageLoad)

        this.timeout = setTimeout(() => {
            this.setState({loaded: true, transitioningImageProps: null}, this.props.onImageLoaded)
        }, this.props.artificialLoadingDelay || 0)
    }

    render() {
        const {
            alt,
            className,
            height,
            hidePlaceholder,
            loadingIndicator,
            placeholderStyle,
            ratio,
            src,
            width,
            useLoaderDuringTransitions,
            itemProp,
            draggable
        } = this.props

        const classes = classNames('pw-image', {
            'pw--is-loaded': this.state.loaded
        }, className)

        const skeletonStyle = {
            ...placeholderStyle,
            height,
            width
        }

        const imageProps = {
            src,
            className: 'pw-image__tag',
            alt,
            height,
            width,
            itemProp,
            draggable
        }

        const loaderNode = (
            <span>
                {loadingIndicator}
                {!hidePlaceholder &&
                    <SkeletonBlock
                        type="div"
                        className="pw--image"
                        style={skeletonStyle}
                    />
                }
            </span>
        )

        // alt is in imageProps
        /* eslint-disable jsx-a11y/img-has-alt */
        const image = this.state.loaded ? (
            <img {...imageProps} />
        ) : (
            <span>
                {useLoaderDuringTransitions || this.state.transitioningImageProps === null ?
                    loaderNode
                :
                    // Image src is changing but we will continue displaying the previous image
                    // until the new one is finished loading, to allow for instant swap
                    <img {...this.state.transitioningImageProps} />
                }
                <img {...imageProps} onLoad={this.imageLoaded} style={{display: 'none'}} />
            </span>
        )

        /* eslint-enable jsx-a11y/img-has-alt */

        return (
            <div className={classes}>
                {ratio ? (
                    <Ratio {...ratio}>
                        {image}
                    </Ratio>
                ) : image}
            </div>
        )
    }
}

Image.defaultProps = {
    alt: '',
    draggable: 'auto',
    src: '',
    placeholderStyle: {},
    onImageLoaded: noop,
    useLoaderDuringTransitions: true
}

Image.propTypes = {
    /**
     * This attribute defines the alternative text describing the image.
     */
    alt: PropTypes.string.isRequired,

    /**
     * This is the image URL.
     */
    src: PropTypes.string.isRequired,

    /**
     * This attribute defines an artificial delay (ms) that slows down when the image is considered loaded.
     */
    artificialLoadingDelay: PropTypes.number,

    /**
     * A CSS class name to be applied to the <img /> element.
     */
    className: PropTypes.string,

    /**
     * Used to determine if the image should be draggable.
     */
    draggable: PropTypes.string,

    /**
     * The intrinsic height of the image in HTML5 CSS pixels, or HTML 4 in pixels or as a percentage.
     */
    height: PropTypes.string,

    /**
    * Used to determine if the image placeholder should be shown while the image loads.
    */
    hidePlaceholder: PropTypes.bool,

    /**
     * A value for the itemprop attribute
     * used to provide microdata to a page for SEO
     * https://www.w3.org/TR/microdata/
     */
    itemProp: PropTypes.string,

    /**
    * Additional content to show with the placeholder while the image loads.
    */
    loadingIndicator: PropTypes.node,

    /**
     * An object whose key is the camelCased version of the style name, and whose value is the style's value, usually a string.
     */
    placeholderStyle: PropTypes.object,

    /**
     * Props for a Ratio component. See the [Ratio](#!/Ratio) component for more details.
     * Example: `{aspect: '4:3'}`
     */
    ratio: PropTypes.object,

    /**
     * Indicates whether to display the loaderNode or the old image when src is changed (while the new image loads)
     */
    useLoaderDuringTransitions: PropTypes.bool,

    /**
     * The intrinsic width of the image in HTML5 CSS pixels, or HTML 4 in pixels or as a percentage.
     */
    width: PropTypes.string,

    /**
     * A callback that gets called when the image is loaded and displayed.
     */
    onImageLoaded: PropTypes.func
}

export default Image
