// progressive-web-sdk/dist/store/products/selectors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getProductThumbnailSrcById = exports.getProductThumbnailById = exports.getProductById = exports.getProductInitialValues = exports.getProductVariants = exports.getProductVariationCategoryIds = exports.getProductVariationCategories = exports.getProductId = exports.getProductPageMeta = exports.getProductCustomContent = exports.getProductHref = exports.getProductAvailability = exports.getProductThumbnail = exports.getProductImages = exports.getProductDescription = exports.getProductPrice = exports.getProductTitle = exports.getCurrentProduct = exports.getCurrentProductId = exports.getProducts = undefined;

var _reselect = require('reselect');

var _immutable = require('immutable');

var _immutable2 = _interopRequireDefault(_immutable);

var _reselectImmutableHelpers = require('reselect-immutable-helpers');

var _selectors = require('../app/selectors');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var getProducts = exports.getProducts = function getProducts(_ref) {
    var products = _ref.products;
    return products;
};
var getCurrentProductId = exports.getCurrentProductId = (0, _reselectImmutableHelpers.createGetSelector)(_selectors.getApp, 'currentProduct');

var getCurrentProduct = exports.getCurrentProduct = (0, _reselectImmutableHelpers.createGetSelector)(getProducts, getCurrentProductId, _immutable2.default.Map());
var getProductTitle = exports.getProductTitle = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentProduct, 'title');
var getProductPrice = exports.getProductPrice = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentProduct, 'price');
var getProductDescription = exports.getProductDescription = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentProduct, 'description');
var getProductImages = exports.getProductImages = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentProduct, 'images', _immutable2.default.List());
var getProductThumbnail = exports.getProductThumbnail = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentProduct, 'thumbnail', _immutable2.default.Map());
var getProductAvailability = exports.getProductAvailability = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentProduct, 'available');
var getProductHref = exports.getProductHref = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentProduct, 'href');
var getProductCustomContent = exports.getProductCustomContent = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentProduct, 'custom', _immutable2.default.Map());
var getProductPageMeta = exports.getProductPageMeta = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentProduct, 'pageMeta', _immutable2.default.Map());
var getProductId = exports.getProductId = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentProduct, 'id');

var getProductVariationCategories = exports.getProductVariationCategories = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentProduct, 'variationCategories', _immutable2.default.List());

var getProductVariationCategoryIds = exports.getProductVariationCategoryIds = (0, _reselect.createSelector)(getProductVariationCategories, function (categories) {
    return categories.map(function (category) {
        return category.get('id');
    });
});
var getProductVariants = exports.getProductVariants = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentProduct, 'variants');
var getProductInitialValues = exports.getProductInitialValues = (0, _reselectImmutableHelpers.createGetSelector)(getCurrentProduct, 'initialValues');

// NOTE: These get-something-By selectors should only be used within actions/commands
// Using them within a component will break the performance optimizations selectors normally give us
var getProductById = exports.getProductById = function getProductById(productId) {
    return (0, _reselectImmutableHelpers.createGetSelector)(getProducts, productId, _immutable2.default.Map());
};
var getProductThumbnailById = exports.getProductThumbnailById = function getProductThumbnailById(id) {
    return (0, _reselectImmutableHelpers.createGetSelector)(getProductById(id), 'thumbnail', _immutable2.default.Map());
};
var getProductThumbnailSrcById = exports.getProductThumbnailSrcById = function getProductThumbnailSrcById(id) {
    return (0, _reselectImmutableHelpers.createGetSelector)(getProductById(id), 'img');
};
