// mobify-integration-manager/dist/integration-manager/api/cart/types.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Cart = exports.CartItems = undefined;

var _runtypes = require('runtypes');

var Runtypes = _interopRequireWildcard(_runtypes);

var _types = require('../types');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2017 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var CartItem = Runtypes.Record({
  id: _types.Identifier,
  productId: _types.ProductID,
  href: _types.URL, // This is temporary
  quantity: _types.Integer,
  /**
   * The unit price of the item
   */
  itemPrice: _types.Money,
  /**
   * The aggregated line price of the item (typically qty * itemPrice)
   */
  linePrice: _types.Money
});

var Coupon = Runtypes.Record({
  id: _types.Identifier,
  couponCode: _types.Text,
  text: _types.Text,
  amount: _types.Money
});

var Tax = Runtypes.Record({
  label: _types.Text,
  amount: _types.Money
});

var Shipping = Runtypes.Record({
  amount: _types.Money
}).And(Runtypes.Partial({
  label: _types.Text
}));

var CartItems = exports.CartItems = Runtypes.Array(CartItem);

var Cart = exports.Cart = Runtypes.Record({
  /**
   * The cart items in the cart
   */
  items: CartItems,
  /**
   * The total of all cart line items before shipping and taxes.
   */
  subtotal: _types.Money,
  /**
   * The total price of the order, including products, shipping and tax
   */
  orderTotal: _types.Money
}).And(Runtypes.Partial({
  /**
   * Page title and meta tag data for SEO purposes
   */
  pageMeta: _types.Meta,
  /**
   * The total discount applied to this cart
   */
  discount: _types.Money,
  /**
   * The coupons and vouchers applied to this cart
   */
  coupons: Runtypes.Array(Coupon),
  /**
   * All taxes applied to items in this cart
   */
  taxes: Tax,
  /**
   * Sum of shipping charges applied to this cart
   */
  shipping: Shipping
}));
