// progressive-web-sdk/dist/components/accordion/accordion-item-content.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Related components:
 *
 * * [AccordionItem](#!/AccordionItem)
 * * [Accordion](#!/Accordion)
 *
 * This component is for internal use only within the AccordionItem.
 *
 * @example ./DESIGN.md
 */

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var AccordionItemContent = function (_React$Component) {
    (0, _inherits3.default)(AccordionItemContent, _React$Component);

    function AccordionItemContent(props) {
        (0, _classCallCheck3.default)(this, AccordionItemContent);

        var _this = (0, _possibleConstructorReturn3.default)(this, (AccordionItemContent.__proto__ || (0, _getPrototypeOf2.default)(AccordionItemContent)).call(this, props));

        _this.state = {
            fullHeight: 0,
            style: {
                maxHeight: '0px',
                transition: 'none'
            }
        };

        _this.animate = _this.animate.bind(_this);
        _this.getFullHeight = _this.getFullHeight.bind(_this);
        return _this;
    }

    (0, _createClass3.default)(AccordionItemContent, [{
        key: 'componentDidAppear',
        value: function componentDidAppear() {
            this.triggerAnimate();
        }
    }, {
        key: 'componentDidEnter',
        value: function componentDidEnter() {
            this.triggerAnimate();
        }
    }, {
        key: 'componentWillLeave',
        value: function componentWillLeave(callback) {
            var _this2 = this;

            var complete = function complete() {
                _this2.props.onAnimationComplete();
                callback();
            };

            this.animate(this.getFullHeight(), 0, complete);
        }
    }, {
        key: 'triggerAnimate',
        value: function triggerAnimate() {
            var _this3 = this;

            var complete = function complete() {
                // After the content has fully opened,
                // remove the set height so the content can grow as needed
                _this3.setState({
                    style: {
                        maxHeight: 'initial',
                        transition: 'none'
                    }
                });

                _this3.props.onAnimationComplete();
            };

            this.animate(0, this.getFullHeight(), complete);
        }
    }, {
        key: 'animate',
        value: function animate(start, end, callback) {
            var _this4 = this;

            var _props = this.props,
                duration = _props.duration,
                easing = _props.easing;


            var frame = function frame() {
                return new _promise2.default(function (resolve) {
                    requestAnimationFrame(resolve);
                });
            };

            var setStartState = function setStartState() {
                return new _promise2.default(function (resolve) {
                    var state = {
                        style: {
                            maxHeight: start + 'px',
                            transition: 'none'
                        }
                    };
                    _this4.setState(state, resolve);
                });
            };

            var setEndState = function setEndState() {
                return new _promise2.default(function (resolve) {
                    var state = {
                        style: {
                            maxHeight: end + 'px',
                            transition: 'max-height ' + duration / 1000 + 's ' + easing
                        }
                    };
                    _this4.setState(state, resolve);
                });
            };

            var waitForAnimation = function waitForAnimation() {
                setTimeout(callback, duration);
            };

            frame().then(setStartState).then(frame).then(setEndState).then(waitForAnimation);
        }
    }, {
        key: 'getFullHeight',
        value: function getFullHeight() {
            return this._content.children[0].offsetHeight;
        }
    }, {
        key: 'render',
        value: function render() {
            var _this5 = this;

            var children = this.props.children;


            return _react2.default.createElement(
                'div',
                { className: 'pw-accordion__content-wrapper',
                    ref: function ref(el) {
                        _this5._content = el;
                    },
                    style: this.state.style
                },
                _react2.default.createElement(
                    'div',
                    { className: 'pw-accordion__content', role: 'presentation' },
                    children
                )
            );
        }
    }]);
    return AccordionItemContent;
}(_react2.default.Component);

AccordionItemContent.propTypes = {
    /**
     * PROVIDED INTERNALLY. Whatever you'd like this AccordionItem to display.
     */
    children: _propTypes2.default.node,

    /**
     * PROVIDED INTERNALLY. Duration of the animation in millis.
     */
    duration: _propTypes2.default.number,

    /**
     * PROVIDED INTERNALLY. Easing function for the animation.
     */
    easing: _propTypes2.default.string,

    /**
     * PROVIDED INTERNALLY. Triggered every time the content has finished animating.
     */
    onAnimationComplete: _propTypes2.default.func
};

exports.default = AccordionItemContent;
