// progressive-web-sdk/dist/routing/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withRouter = exports.browserHistory = exports.IndexRoute = exports.IndexRedirect = exports.Route = exports.QueryRoute = exports.QueryRouter = exports.SelectorRouter = exports.SelectorRoute = exports.Router = undefined;

var _router = require('./router');

var _router2 = _interopRequireDefault(_router);

var _selectorRoute = require('./selector-route');

var _queryRouter = require('./query-router');

var _reactRouter = require('react-router');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

exports.Router = _router2.default;
exports.SelectorRoute = _selectorRoute.SelectorRoute;
exports.SelectorRouter = _selectorRoute.SelectorRouter;
exports.QueryRouter = _queryRouter.QueryRouter;
exports.QueryRoute = _queryRouter.QueryRoute;
exports.Route = _reactRouter.Route;
exports.IndexRedirect = _reactRouter.IndexRedirect;
exports.IndexRoute = _reactRouter.IndexRoute;
exports.browserHistory = _reactRouter.browserHistory;
exports.withRouter = _reactRouter.withRouter;
