// progressive-web-sdk/dist/components/bazaarvoice-wrapper/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _componentUtils = require('../../utils/component-utils');

var _error = require('../../analytics/error');

var _error2 = _interopRequireDefault(_error);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * This Higher Order Component can be used to construct
 * components that use Bazaarvoice.
 * It loads the Bazaarvoice API script once
 * and renders its wrapped component after the script has been loaded.
 * This means that the wrapped components can use componentDidMount
 * to run any initialization code that relies on the api script.
 */

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var scriptClass = 'js-bvapi';
var loadingClass = 'pw--loading';
var loadedClass = 'pw--loaded';

var bazaarvoiceWrapper = function bazaarvoiceWrapper(WrappedComponent) {
    var BazaarvoiceWrapper = function (_React$Component) {
        (0, _inherits3.default)(BazaarvoiceWrapper, _React$Component);

        function BazaarvoiceWrapper(props) {
            (0, _classCallCheck3.default)(this, BazaarvoiceWrapper);

            var _this = (0, _possibleConstructorReturn3.default)(this, (BazaarvoiceWrapper.__proto__ || (0, _getPrototypeOf2.default)(BazaarvoiceWrapper)).call(this, props));

            _this.state = {
                apiLoaded: false
            };

            _this.WrappedComponent = WrappedComponent;
            return _this;
        }

        (0, _createClass3.default)(BazaarvoiceWrapper, [{
            key: 'componentDidMount',
            value: function componentDidMount() {
                var _this2 = this;

                var _props = this.props,
                    apiSrc = _props.apiSrc,
                    configurationOptions = _props.configurationOptions,
                    onError = _props.onError;


                var existingScript = document.querySelector('.' + scriptClass);

                if (!existingScript) {
                    var script = document.createElement('script');
                    script.src = apiSrc;
                    script.className = scriptClass + ' ' + loadingClass;
                    script.onload = function () {
                        _this2.setState({
                            apiLoaded: true
                        });

                        window.$BV.configure('global', configurationOptions);
                        script.classList.remove(loadingClass);
                        script.classList.add(loadedClass);
                    };
                    script.onerror = function () {
                        onError();

                        try {
                            // This AppError automatically sends analytics about this error to our Engagement Engine
                            // However, we want to immediately catch it because we don't want to break the PWA
                            // (or make anyone looking at the console think we broke the PWA)
                            // just because a third party plugin failed to load
                            throw new _error2.default('Bazarrvoice Error');
                        } catch (e) {
                            console.error(e);
                        }
                    };

                    document.body.appendChild(script);
                } else if (new RegExp(loadedClass).test(existingScript.className)) {
                    // Another instance has already loaded this script
                    this.setState({
                        apiLoaded: true
                    });
                }
            }
        }, {
            key: 'render',
            value: function render() {
                return this.state.apiLoaded ? _react2.default.createElement(WrappedComponent, this.props) : false;
            }
        }]);
        return BazaarvoiceWrapper;
    }(_react2.default.Component);

    BazaarvoiceWrapper.WrappedComponent = WrappedComponent;
    BazaarvoiceWrapper.displayName = (0, _componentUtils.getDisplayName)(WrappedComponent);
    BazaarvoiceWrapper.defaultProps = {
        configurationOptions: {},
        onError: function onError() {}
    };
    BazaarvoiceWrapper.propTypes = {
        apiSrc: _propTypes2.default.string,
        configurationOptions: _propTypes2.default.object,
        onError: _propTypes2.default.func
    };

    return BazaarvoiceWrapper;
};

exports.default = bazaarvoiceWrapper;
