// progressive-web-sdk/dist/integration-manager/account/results.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.receiveSavedAddresses = exports.receiveAccountAddressUIData = exports.receiveWishlistCustomContent = exports.receiveUpdatedWishlistItem = exports.removeWishlistItem = exports.receiveWishlistUIData = exports.receiveWishlistData = exports.receiveCurrentOrderNumber = exports.receiveAccountOrderListData = exports.receiveAccountOrderUIData = exports.receiveAccountInfoUIData = exports.receiveAccountInfoData = exports.receiveAccountDashboardUIData = exports.receiveRegisterData = exports.setRegisterLoaded = exports.receiveSigninData = exports.setSigninLoaded = undefined;

var _actionCreation = require('../../utils/action-creation');

var setSigninLoaded = exports.setSigninLoaded = (0, _actionCreation.createAction)('Set Sign In Page Loaded'); /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var receiveSigninData = exports.receiveSigninData = (0, _actionCreation.createAction)('Receive Sign In Page Data');

var setRegisterLoaded = exports.setRegisterLoaded = (0, _actionCreation.createAction)('Set Register Page Loaded');
var receiveRegisterData = exports.receiveRegisterData = (0, _actionCreation.createAction)('Receive Register Page Data');

var receiveAccountDashboardUIData = exports.receiveAccountDashboardUIData = (0, _actionCreation.createAction)('Receive Account Dashboard Data');

var receiveAccountInfoData = exports.receiveAccountInfoData = (0, _actionCreation.createAction)('Receive Account Info Data');
var receiveAccountInfoUIData = exports.receiveAccountInfoUIData = (0, _actionCreation.createAction)('Receive Account Info UI Data');

var receiveAccountOrderUIData = exports.receiveAccountOrderUIData = (0, _actionCreation.createAction)('Receive Account Order UI Data');
var receiveAccountOrderListData = exports.receiveAccountOrderListData = (0, _actionCreation.createAction)('Receive Account Order List Data', ['orders']);

var receiveCurrentOrderNumber = exports.receiveCurrentOrderNumber = (0, _actionCreation.createAction)('Receive Current Order Id', ['currentOrderNumber']);

var receiveWishlistData = exports.receiveWishlistData = (0, _actionCreation.createAction)('Receive wishlist data', ['wishlist']);
var receiveWishlistUIData = exports.receiveWishlistUIData = (0, _actionCreation.createAction)('Receive Wishlist UI Data');
var removeWishlistItem = exports.removeWishlistItem = (0, _actionCreation.createAction)('Remove Wishlist Item', ['removeItemId']);
var receiveUpdatedWishlistItem = exports.receiveUpdatedWishlistItem = (0, _actionCreation.createAction)('Receives a updated Wishlist Item data');
var receiveWishlistCustomContent = exports.receiveWishlistCustomContent = (0, _actionCreation.createAction)('Receive Wishlist custom content');

var receiveAccountAddressUIData = exports.receiveAccountAddressUIData = (0, _actionCreation.createAction)('Receive Account Address UI Data');
var receiveSavedAddresses = exports.receiveSavedAddresses = (0, _actionCreation.createAction)('Receive Saved Addresses', ['savedAddresses']);
