// progressive-web-sdk/src/components/bazaarvoice-review/index.jsx
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import bazaarvoiceWrapper from '../bazaarvoice-wrapper'

/**
 * This component makes it easier to integrate Bazaarvoice reviews into your PWA.
 *  This component allows you to render the reviews for a product based on a product ID.
 */

class BazaarvoiceReview extends React.Component {
    componentDidMount() {
        const {
            productId
        } = this.props

        const $BV = window.$BV

        // We can't really test that a function that doesn't exist wasn't called
        /* istanbul ignore else */
        if ($BV) {
            $BV.ui('rr', 'show_reviews', {
                productId
            })
        }
    }

    render() {
        const {
            className,
            containerId
        } = this.props

        const classes = classNames('pw-bazaarvoice-review', className)

        return (
            <div className={classes}>
                <div id={containerId} />
            </div>
        )
    }
}


BazaarvoiceReview.defaultProps = {
    containerId: 'BVRRContainer'
}

BazaarvoiceReview.propTypes = {
    /**
     * The src for the Bazaarvoice API script
     */
    apiSrc: PropTypes.string.isRequired,

    /**
     * The product ID the reviews should be shown for
     */
    productId: PropTypes.string.isRequired,

    /**
     * Adds values to the `class` attribute of the root element
     */
    className: PropTypes.string,

    /**
     * This id is added to the element Bazaarvoice should render the
     * review content in
     */
    containerId: PropTypes.string

}

export default bazaarvoiceWrapper(BazaarvoiceReview)
