// progressive-web-sdk/dist/components/sheet/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _TransitionGroup = require('react-transition-group/TransitionGroup');

var _TransitionGroup2 = _interopRequireDefault(_TransitionGroup);

var _sheetContent = require('./sheet-content');

var _sheetContent2 = _interopRequireDefault(_sheetContent);

var _propTypeUtils = require('../../prop-type-utils');

var _utils = require('../../utils/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var TRANSITION_GROUP_STYLE = {
  zIndex: 1000,
  position: 'fixed',
  top: 0,
  left: 0,
  width: 0,
  height: 0

  /**
   * Sheets are modal containers that animate into view from one side of the screen
   * to reveal more content. Typically, sheets are only displayed as a result of
   * user initiated action, and persist until dismissed. Common uses for sheet
   * include, but not limited to, dialog modals and sidebar navigations.
   *
   * @example ./DESIGN.md
   */
};
var Sheet = function (_React$Component) {
  (0, _inherits3.default)(Sheet, _React$Component);

  function Sheet(props) {
    (0, _classCallCheck3.default)(this, Sheet);

    var _this = (0, _possibleConstructorReturn3.default)(this, (Sheet.__proto__ || (0, _getPrototypeOf2.default)(Sheet)).call(this, props));

    _this.bodyEl = null;
    return _this;
  }

  (0, _createClass3.default)(Sheet, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      this.bodyEl = document.body;
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          open = _props.open,
          prerender = _props.prerender,
          children = _props.children,
          className = _props.className;


      return _react2.default.createElement(
        'div',
        { className: 'pw-sheet__outer-wrapper' },
        _react2.default.createElement(
          _TransitionGroup2.default,
          { component: 'div', style: TRANSITION_GROUP_STYLE },
          open ? _react2.default.createElement(_sheetContent2.default, this.props) : null
        ),
        !open && prerender && _react2.default.createElement(
          'div',
          {
            className: 'pw-sheet__prerendered-children u-visually-hidden ' + className,
            'aria-hidden': 'true'
          },
          children
        )
      );
    }
  }]);
  return Sheet;
}(_react2.default.Component);

Sheet.propTypes = {
  /**
   * User-defined content of the sheet, can be a combination of markup and/or
   * React components.
   */
  children: _propTypes2.default.node,

  /**
   * Adds values to the `class` attribute of the root element.
   */
  className: _propTypes2.default.string,

  /**
   * Coverage defines the amount of the current viewport taken up by the
   * sheet content.
   */
  coverage: _propTypeUtils.percentage,

  /**
   * Duration will define the time the animation takes to complete.
   */
  duration: _propTypes2.default.number,

  /**
   * Easing function for the animation as a string whose value
   * can be anything the [CSS `transition-timing-function`](https://developer.mozilla.org/en/docs/Web/CSS/transition-timing-function)
   * can take in.
   */
  easing: _propTypes2.default.string,

  /**
   * Type will define the entry point of the sheet. See below for details.
   */
  effect: _propTypes2.default.oneOf(['slide-top', 'slide-right', 'slide-bottom', 'slide-left', 'modal-center']),

  /**
   * User-defined footer of the sheet.
   */
  footerContent: _propTypes2.default.element,

  /**
   * User-defined header of the sheet.
   */
  headerContent: _propTypes2.default.element,

  /**
   * Set ID of root element. Generally used to backtrace the source of a
   * sheet.
   */
  id: _propTypes2.default.string,

  /**
   * Level of opacity on sheet mask.
   */
  maskOpacity: _propTypes2.default.number,

  /**
   * Determines whether the Sheet is opened or closed.
   */
  open: _propTypes2.default.bool,

  /**
   * This is used by Sheet to prerender the content before openning it.
   */
  prerender: _propTypes2.default.bool,

  /**
   * Dictate whether the height of the modal grows to the maximum height as
   * declared by the `coverage` prop, or to shrink down to the height of the
   * modal's content.
   */
  shrinkToContent: _propTypes2.default.bool,

  /**
   * User-defined title of the sheet.
   */
  title: _propTypes2.default.string,

  // Callback props

  /**
   * User-defined function that is called before the sheet is closed.
   */
  onBeforeClose: _propTypes2.default.func,

  /**
   * User-defined function that is called before the sheet is opened.
   */
  onBeforeOpen: _propTypes2.default.func,

  /**
   * User-defined function that is called after the sheet is closed.
   */
  onClose: _propTypes2.default.func,

  /**
   * User-defined function that is called when the user clicks to dismiss the
   * sheet. Use this to change state and re-render the Sheet through props.
   */
  onDismiss: _propTypes2.default.func,

  /**
   * User-defined function that is called after the sheet is opened.
   */
  onOpen: _propTypes2.default.func

  /* istanbul ignore next */

};Sheet.defaultProps = {
  coverage: '80%',
  duration: 200,
  onBeforeClose: _utils.noop,
  onBeforeOpen: _utils.noop,
  onClose: _utils.noop,
  onOpen: _utils.noop,
  onDismiss: _utils.noop,
  maskOpacity: 0.5,
  open: false,
  easing: 'ease-in-out',
  effect: 'slide-left',
  shrinkToContent: false
};

exports.default = Sheet;
