// progressive-web-sdk/dist/components/expiry-date/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _cardUtils = require('../../card-utils');

var _dataObjects = require('../../analytics/data-objects/');

var _reactInputMask = require('react-input-mask');

var _reactInputMask2 = _interopRequireDefault(_reactInputMask);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * An input field to capture credit card's expiry dates
 *
 * @example ./DESIGN.md
 */

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var ExpiryDate = function ExpiryDate(_ref) {
  var className = _ref.className,
      placeholder = _ref.placeholder,
      defaultValue = _ref.defaultValue,
      onChange = _ref.onChange,
      onBlur = _ref.onBlur;

  var classes = (0, _classnames2.default)('pw-expiry-date', className);

  return _react2.default.createElement(
    'div',
    { className: classes },
    _react2.default.createElement(_reactInputMask2.default, {
      type: 'tel',
      placeholder: placeholder,
      defaultValue: defaultValue,
      onChange: (0, _cardUtils.toStoreValueDecorator)(onChange, _cardUtils.expiryFormat),
      onBlur: (0, _cardUtils.toStoreValueDecorator)(onBlur, _cardUtils.expiryFormat),
      mask: '99/99',
      maskChar: null,
      'data-analytics-name': _dataObjects.UI_NAME.cardExpiryDate
    })
  );
};

ExpiryDate.propTypes = {
  /**
   * Adds values to the `class` attribute of the root element.
   */
  className: _propTypes2.default.string,

  /**
   * Adds a default value to input
   */
  defaultValue: _propTypes2.default.string,

  /**
   * Adds a placeholder.
   */
  placeholder: _propTypes2.default.string,

  /**
   * PROVIDED INTERNALLY: OnBlur callback with signature `(eventOrValue) => undefined`
   * (passed in by redux-form).
   */
  onBlur: _propTypes2.default.func,

  /**
   * PROVIDED INTERNALLY: OnChange callback with signature `(eventOrValue) => undefined`
   * (passed in by redux-form).
   */
  onChange: _propTypes2.default.func

};

exports.default = ExpiryDate;
