// progressive-web-sdk/scripts/raml/generate-record.test.js
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

const generateRecord = require('./generate-record')

const testRecord = {
    name: 'Test',
    contents: [
        {
            name: 'item',
            type: 'number',
            defaultValue: '0',
            propType: 'PropTypes.number'
        },
        {
            name: 'subrecord',
            type: 'SubRecord',
            defaultValue: 'new SubRecord()',
            propType: 'PropTypes.instanceOf(SubRecord)',
            importVars: 'SubRecord, {SubRecord}',
            custom: true
        }
    ]
}

const testOutput = [
    'import Immutable from \'immutable\'',
    'import PropTypes from \'prop-types\'',
    'import SubRecord, {SubRecord} from \'./sub-record\'',
    '',
    'const Test = Immutable.Record({',
    '    item: 0,',
    '    subrecord: new SubRecord()',
    '})',
    '',
    'export const TestPropType = PropTypes.shape({',
    '    item: PropTypes.number,',
    '    subrecord: PropTypes.instanceOf(SubRecord)',
    '})',
    '',
    'export default Test\n'
].join('\n')

test('generates correct JS', () => {
    expect(generateRecord(testRecord)).toEqual({
        filename: 'test.js',
        text: testOutput
    })
})
