// mobify-integration-manager/dist/connectors/sfcc/checkout/parsers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.parseShippingAddressFromBasket = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2017 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

/* eslint-disable camelcase */
var parseShippingAddressFromBasket = exports.parseShippingAddressFromBasket = function parseShippingAddressFromBasket(basketData) {
    var email = basketData.customer_info.email,
        _basketData$shipments = (0, _slicedToArray3.default)(basketData.shipments, 1),
        shipping_address = _basketData$shipments[0].shipping_address;

    var initialValues = {
        countryId: 'us',
        firstname: '',
        lastname: '',
        addressLine1: '',
        city: '',
        postcode: '',
        telephone: ''
    };

    if (shipping_address) {
        initialValues = (0, _extends3.default)({}, initialValues, {
            username: email,
            name: shipping_address.full_name,
            firstname: shipping_address.first_name,
            lastname: shipping_address.last_name,
            company: shipping_address.company_name,
            addressLine1: shipping_address.address1,
            addressLine2: shipping_address.address2,
            countryId: shipping_address.country_code,
            city: shipping_address.city,
            regionId: shipping_address.state_code,
            postcode: shipping_address.postal_code,
            telephone: shipping_address.phone
        });
    }

    return initialValues;
};
/* eslint-enable camelcase */
