// mobify-integration-manager/dist/connectors/sfcc/checkout/utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.createOrderAddressObject = exports.populateLocationsData = undefined;

var _constants = require('./constants');

var _results = require('../../../integration-manager/api/checkout/results');

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2017 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var populateLocationsData = exports.populateLocationsData = function populateLocationsData() {
    return (0, _results.receiveCheckoutLocations)({
        countries: [{
            id: '',
            label: '',
            regionRequired: false,
            postcodeRequired: false
        }, {
            id: 'us',
            label: 'United States',
            regionRequired: true,
            postcodeRequired: true
        }],
        regions: _constants.STATES
    });
};

var createOrderAddressObject = exports.createOrderAddressObject = function createOrderAddressObject(formValues) {
    var name = formValues.name,
        addressName = formValues.addressName,
        firstname = formValues.firstname,
        lastname = formValues.lastname,
        company = formValues.company,
        addressLine1 = formValues.addressLine1,
        addressLine2 = formValues.addressLine2,
        countryId = formValues.countryId,
        city = formValues.city,
        preferred = formValues.preferred,
        regionId = formValues.regionId,
        postcode = formValues.postcode,
        telephone = formValues.telephone;


    return {
        address1: addressLine1,
        address2: addressLine2,
        address_id: addressName,
        city: city,
        country_code: countryId,
        first_name: firstname,
        last_name: lastname,
        full_name: name,
        phone: telephone,
        preferred: preferred,
        postal_code: postcode,
        state_code: regionId,
        company_name: company
    };
};
