// progressive-web-sdk/src/components/rating/index.jsx
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import Icon from '../icon'
import Image from '../image'

/**
 * Rating is used to display an average user satisfaction level of a product, it
 * is the summary of the ratings given by users in the product reviews section.
 *
 * @example ./DESIGN.md
 */

const RatingIcon = ({
    name,
    src,
    modifierClass
}) => {
    const iconClasses = classNames('pw-rating__icon', modifierClass)

    return (
        <div className={iconClasses}>
            {src ?
                <Image role="presentation" src={src} />
            :
                <Icon name={name} />
            }
        </div>
    )
}

RatingIcon.propTypes = {
    /**
     * PROVIDED INTERNALLY: The addtional classes to be used for the RatingIcon div
     */
    modifierClass: PropTypes.string,
    /**
     * PROVIDED INTERNALLY: The name of the icon to be used
     */
    name: PropTypes.string,
    /**
     * PROVIDED INTERNALLY: The URL of the image to be used. Icon is used otherwise.
     */
    src: PropTypes.string
}

RatingIcon.displayName = 'RatingIcon'

/**
 * Commonly used to display user satisfaction of products or services.
 */
const Rating = ({
    className,
    count,
    total,
    name,
    src
}) => {
    const ratingPercentage = count / total * 100
    const classes = classNames('pw-rating', className)

    return (
        <div className={classes}>
            <div className="pw-rating__label">
                Rating is {count} out of {total}
            </div>

            <div
                className="pw-rating__icon-wrapper"
                role="presentation"
                aria-hidden="true"
            >
                <div
                    className="pw-rating__filled-icons"
                    style={{width: `${ratingPercentage}%`}}
                >
                    {Array(total).fill()
                        .map((item, i) => (
                            <RatingIcon
                                key={i}
                                name={name}
                                src={src}
                                modifierClass="pw--filled"
                            />
                        )
                    )}
                </div>

                {Array(total).fill()
                    .map((item, i) => (
                        <RatingIcon key={i} name={name} src={src} />
                    )
                )}
            </div>
        </div>
    )
}

Rating.defaultProps = {
    count: 0,
    total: 5,
    name: 'star'
}

Rating.propTypes = {
    /**
     * Adds values to the `class` attribute of the root element.
     */
    className: PropTypes.string,

    /**
     * The number of filled stars to render (can be a fraction).
     */
    count: PropTypes.number,

    /**
     * Name of the SVG sprite you want to use for the rating star.
     * Not used if the src prop is set.
     */
    name: PropTypes.string,

    /**
     * URL for the image icon to use for the rating star.
     * If not provided, the SVG icon described by the `name` prop is used.
     */
    src: PropTypes.string,

    /**
     * The total number of stars to render.
     */
    total: PropTypes.number
}

export default Rating
