// progressive-web-sdk/dist/components/stepper/partials/stepper-button.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _button = require('../../button');

var _button2 = _interopRequireDefault(_button);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var StepperButton = function StepperButton(_ref) {
  var className = _ref.className,
      analyticsName = _ref.analyticsName,
      disabled = _ref.disabled,
      icon = _ref.icon,
      text = _ref.text,
      title = _ref.title,
      onClick = _ref.onClick,
      isDecrement = _ref.isDecrement;

  var buttonClass = isDecrement ? 'pw--decrement' : 'pw--increment';
  var classes = (0, _classnames2.default)('pw-stepper__button', className, '' + buttonClass);
  return _react2.default.createElement(
    'div',
    { className: 'pw-stepper__action' },
    _react2.default.createElement(_button2.default, {
      className: classes,
      icon: icon,
      text: text,
      title: title,
      disabled: disabled,
      'data-analytics-name': analyticsName,
      onClick: onClick
    })
  );
};

StepperButton.propTypes = {
  /**
   * Analytics name to use for the button.
   */
  analyticsName: _propTypes2.default.string,
  /**
   * Adds values to the `class` attribute of the root element.
   */
  className: _propTypes2.default.string,
  /**
   * Defines if the button is disabled.
   */
  disabled: _propTypes2.default.bool,
  /**
   * Icon to use for the button.
   */
  icon: _propTypes2.default.string,
  /**
   * Determines whether it is a decrement button
   */
  isDecrement: _propTypes2.default.bool,
  /**
   * Text for the button.
   */
  text: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node]),
  /**
   * Title for the button.
   */
  title: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.node]),
  /**
   * Click handler for the button.
   */
  onClick: _propTypes2.default.func
};

exports.default = StepperButton;
