// progressive-web-sdk/dist/analytics/connectors/google-analytics/mobify-ga.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _get2 = require('babel-runtime/helpers/get');

var _get3 = _interopRequireDefault(_get2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _connector = require('./connector');

var _connector2 = _interopRequireDefault(_connector);

var _dataObjects = require('../../data-objects/');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var ga = void 0;

var cleanProduct = function cleanProduct(product) {
    return product ? new _dataObjects.Product(product, [], {}, false) : undefined;
};

var MobifyGA = function (_GoogleAnalytics) {
    (0, _inherits3.default)(MobifyGA, _GoogleAnalytics);

    function MobifyGA(options) {
        (0, _classCallCheck3.default)(this, MobifyGA);

        var _this = (0, _possibleConstructorReturn3.default)(this, (MobifyGA.__proto__ || (0, _getPrototypeOf2.default)(MobifyGA)).call(this, '🐯 Mobify Google Analytics', 'mobifyTracker', options));

        if (!_this.options.mobifyGAID) {
            throw new Error('options cannot be undefined and must have "mobifyGAID" property defined');
        }
        return _this;
    }

    (0, _createClass3.default)(MobifyGA, [{
        key: 'ready',
        value: function ready() {
            var _this2 = this;

            ga = window.ga;

            ga('create', this.options.mobifyGAID, 'auto', { name: this.trackerName });

            (0, _get3.default)(MobifyGA.prototype.__proto__ || (0, _getPrototypeOf2.default)(MobifyGA.prototype), 'ready', this).call(this, function () {
                _this2.setDimension(1, '1');
            });
        }
    }, {
        key: 'debugCallback',
        value: function debugCallback(type) {
            switch (type) {
                case _dataObjects.EVENT_ACTION.pageview:
                    {
                        console.log('Template Name:\t\t ' + this.tracker.get('dimension2'));
                        break;
                    }
            }

            console.groupEnd();
        }
    }, {
        key: 'pageviewEvent',
        value: function pageviewEvent(payload) {
            var page = new _dataObjects.Page(payload, [_dataObjects.Page.TEMPLATENAME], {}, false);

            this.setDimension(2, page.templateName);

            return (0, _get3.default)(MobifyGA.prototype.__proto__ || (0, _getPrototypeOf2.default)(MobifyGA.prototype), 'pageviewEvent', this).call(this, payload);
        }
    }, {
        key: 'addToCartEvent',
        value: function addToCartEvent(payload) {
            this.ecAddProductAction('add', cleanProduct(payload.product));

            return ['send', 'event', 'UX', 'click', 'add to cart'];
        }
    }, {
        key: 'removeFromCartEvent',
        value: function removeFromCartEvent(payload) {
            this.ecAddProductAction('remove', cleanProduct(payload.product));

            return ['send', 'event', 'UX', 'click', 'remove from cart'];
        }
    }]);
    return MobifyGA;
}(_connector2.default);

exports.default = MobifyGA;
