// progressive-web-sdk/dist/components/skeleton-block/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Related components:
 *
 * * [SkeletonInline](#!/SkeletonInline)
 * * [SkeletonText](#!/SkeletonText)
 *
 * A skeleton component that can be used to mimic block-level components, such as images.
 *
 * @example ./DESIGN.md
 */
var SkeletonBlock = function SkeletonBlock(props) {
    var classes = (0, _classnames2.default)('pw-skeleton-block', props.className);

    var styles = (0, _extends3.default)({
        height: props.height,
        width: props.width
    }, props.style);

    var attributes = (0, _extends3.default)({}, props, {
        role: 'presentation',
        className: classes,
        style: styles
    });

    return _react2.default.createElement(props.type, attributes);
}; /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

SkeletonBlock.defaultProps = {
    height: 'auto',
    style: {},
    type: 'div',
    width: '100%'
};

SkeletonBlock.propTypes = {
    /**
    * Adds values to the `class` attribute of the root element.
    */
    className: _propTypes2.default.string,

    /**
    * Specifies the height of the skeleton, in any valid dimension type.
    */
    height: _propTypes2.default.string,

    /**
    * Specifies any inline styles to be applied to the element.
    */
    style: _propTypes2.default.object,

    /**
    * Specifies the element type to be constructed.
    */
    type: _propTypes2.default.string,

    /**
    * Specifies the width of the skeleton's children's width, in any valid dimension type.
    */
    width: _propTypes2.default.string
};

exports.default = SkeletonBlock;
