// mobify-integration-manager/dist/connectors/hybris/app/parser.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.parseCurrencies = exports.parseSearchSuggestions = undefined;

var _config = require('../config');

var parseSearchSuggestions = exports.parseSearchSuggestions = function parseSearchSuggestions(_ref) {
    var suggestions = _ref.suggestions;

    var searchSuggestions = null;

    if (suggestions.length) {
        searchSuggestions = suggestions.map(function (data) {
            var searchTerm = data.value;
            return {
                href: (0, _config.buildSearchURL)(searchTerm),
                children: searchTerm,
                endAction: ''
            };
        });
    }

    return searchSuggestions;
}; /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2017 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
var parseCurrencies = exports.parseCurrencies = function parseCurrencies(currencies) {
    return currencies.map(function (currency) {
        return {
            label: currency.name,
            symbol: currency.symbol,
            code: currency.isocode
        };
    });
};
