// progressive-web-sdk/dist/store/user/selectors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getWishlistItemCount = exports.getWishlistProducts = exports.getWishlistCustomContent = exports.getWishlistItems = exports.getWishlistShareURL = exports.getWishlistID = exports.getWishlistTitle = exports.getWishlist = exports.getAddresses = exports.getDefaultAddress = exports.getSavedAddresses = exports.getUserCustomContent = exports.getIsLoggedIn = exports.getUser = undefined;

var _immutable = require('immutable');

var _immutable2 = _interopRequireDefault(_immutable);

var _reselect = require('reselect');

var _reselectImmutableHelpers = require('reselect-immutable-helpers');

var _selectors = require('../../store/products/selectors');

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var getUser = exports.getUser = function getUser(_ref) {
    var user = _ref.user;
    return user;
}; // /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
// /* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
// /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
var getIsLoggedIn = exports.getIsLoggedIn = (0, _reselectImmutableHelpers.createGetSelector)(getUser, 'isLoggedIn');

var getUserCustomContent = exports.getUserCustomContent = (0, _reselectImmutableHelpers.createGetSelector)(getUser, 'custom', _immutable2.default.Map());

var getSavedAddresses = exports.getSavedAddresses = (0, _reselectImmutableHelpers.createGetSelector)(getUser, 'savedAddresses', _immutable2.default.List());

var PLACEHOLDER = {
    text: undefined
};

var getDefaultAddress = exports.getDefaultAddress = (0, _reselect.createSelector)(getUser, function (user) {
    var addresses = user.get('savedAddresses');
    return addresses ? addresses.toJS().find(function (address) {
        return address.preferred;
    }) : _immutable2.default.Map();
});

var getAddresses = exports.getAddresses = (0, _reselect.createSelector)(getUser, function (user) {
    var addresses = user.get('savedAddresses');
    return addresses ? addresses.toJS().filter(function (address) {
        return !address.preferred;
    }) : _immutable2.default.List(new Array(5).fill(PLACEHOLDER));
});

var getWishlist = exports.getWishlist = (0, _reselectImmutableHelpers.createGetSelector)(getUser, 'wishlist', _immutable2.default.Map());

var getWishlistTitle = exports.getWishlistTitle = (0, _reselectImmutableHelpers.createGetSelector)(getWishlist, 'title', _constants.WISHLIST_TITLE_DEFAULT);

var getWishlistID = exports.getWishlistID = (0, _reselectImmutableHelpers.createGetSelector)(getWishlist, 'id', '');

var getWishlistShareURL = exports.getWishlistShareURL = (0, _reselectImmutableHelpers.createGetSelector)(getWishlist, 'shareURL');

var getWishlistItems = exports.getWishlistItems = (0, _reselectImmutableHelpers.createGetSelector)(getWishlist, 'products', _immutable2.default.List());

var getWishlistCustomContent = exports.getWishlistCustomContent = (0, _reselectImmutableHelpers.createGetSelector)(getWishlist, 'custom', _immutable2.default.Map());

var getWishlistProducts = exports.getWishlistProducts = (0, _reselect.createSelector)(getWishlistItems, _selectors.getProducts, function (wishlistItems, products) {
    return wishlistItems.map(function (wishlistItem) {
        var productData = products.get(wishlistItem.get('productId'));
        if (productData) {
            return productData.mergeDeep(wishlistItem);
        }
        return _immutable2.default.Map();
    });
});

var getWishlistItemCount = exports.getWishlistItemCount = (0, _reselect.createSelector)(getWishlistItems, function (products) {
    return products.size;
});
