// progressive-web-sdk/src/store/app/selectors.js
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

import Immutable from 'immutable'
import {createSelector} from 'reselect'
import {createGetSelector} from 'reselect-immutable-helpers'
import {urlToPathKey, urlToBasicPathKey} from '../../utils/utils'
import {CURRENT_URL} from './constants'

export const getApp = ({app}) => app
export const getCurrentUrl = createGetSelector(getApp, CURRENT_URL)
export const getCurrentPathKey = createSelector(getCurrentUrl, urlToPathKey)
export const getCurrentPathKeyWithoutQuery = createSelector(getCurrentPathKey, urlToBasicPathKey)
export const isStandaloneApp = createGetSelector(getApp, 'standaloneApp')
export const getSelectedCurrency = createGetSelector(getApp, 'selectedCurrency', {})
export const getAvailableCurrencies = createGetSelector(getApp, 'availableCurrencies', Immutable.List())
