// progressive-web-sdk/dist/components/expose-api-base/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _routing = require('../../routing');

var _isReactRoute = require('../../routing/is-react-route');

var _isReactRoute2 = _interopRequireDefault(_isReactRoute);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Base PWA methods exposed to **window.Progressive.api** namescope
 *
 * 1. **navigate(pathOrUrl)**
 *
 *      Invokes PWA navigation
 *
 *      **Usage example:**
 *
 *          window.Progressive.api.navigate('/');
 **/
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var ExposeApiBase = function (_Component) {
    (0, _inherits3.default)(ExposeApiBase, _Component);

    function ExposeApiBase() {
        (0, _classCallCheck3.default)(this, ExposeApiBase);
        return (0, _possibleConstructorReturn3.default)(this, (ExposeApiBase.__proto__ || (0, _getPrototypeOf2.default)(ExposeApiBase)).apply(this, arguments));
    }

    (0, _createClass3.default)(ExposeApiBase, [{
        key: 'componentDidMount',
        value: function componentDidMount() {
            if (window.Progressive && window.Progressive.api) {
                return;
            }

            window.Progressive = window.Progressive || {};
            window.Progressive.api = this.buildProgressiveApi();
        }
    }, {
        key: 'buildProgressiveApi',
        value: function buildProgressiveApi() {
            return {
                navigate: function navigate(url) {
                    /* istanbul ignore next */
                    if ((0, _isReactRoute2.default)(url)) {
                        _routing.browserHistory.push({ pathname: url });
                    } else {
                        window.location.href = url;
                    }
                }
            };
        }
    }, {
        key: 'render',
        value: function render() {
            return null;
        }
    }]);
    return ExposeApiBase;
}(_react.Component);

// This is defined so that styleguidist can auto generate doc for this component


ExposeApiBase.propTypes = {
    /**
     * This component doesn't take any props
     */
    none: _propTypes2.default.node
};

exports.default = ExposeApiBase;
