// progressive-web-sdk/dist/store/offline/actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.clearOfflinePages = exports.trackOfflinePage = exports.clearOfflineModeStartTime = exports.setOfflineModeStartTime = exports.setFetchedPage = exports.clearPageFetchError = exports.setPageFetchError = undefined;

var _actionCreation = require('../../utils/action-creation');

var setPageFetchError = exports.setPageFetchError = (0, _actionCreation.createAction)('Set page fetch error', ['fetchError']); /* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var clearPageFetchError = exports.clearPageFetchError = (0, _actionCreation.createAction)('Clear page fetch error');

var setFetchedPage = exports.setFetchedPage = (0, _actionCreation.createAction)('Set fetched page', ['url']);

var setOfflineModeStartTime = exports.setOfflineModeStartTime = (0, _actionCreation.createAction)('Set start time for offline mode', ['startTime']);

var clearOfflineModeStartTime = exports.clearOfflineModeStartTime = (0, _actionCreation.createAction)('Clear offline mode start time');

var trackOfflinePage = exports.trackOfflinePage = (0, _actionCreation.createAction)('Track Offline Page viewed');

var clearOfflinePages = exports.clearOfflinePages = (0, _actionCreation.createAction)('Clear Offline Pages');
