// progressive-web-sdk/dist/components/bazaarvoice-review/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _bazaarvoiceWrapper = require('../bazaarvoice-wrapper');

var _bazaarvoiceWrapper2 = _interopRequireDefault(_bazaarvoiceWrapper);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * This component makes it easier to integrate Bazaarvoice reviews into your PWA.
 *  This component allows you to render the reviews for a product based on a product ID.
 */

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2018 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var BazaarvoiceReview = function (_React$Component) {
    (0, _inherits3.default)(BazaarvoiceReview, _React$Component);

    function BazaarvoiceReview() {
        (0, _classCallCheck3.default)(this, BazaarvoiceReview);
        return (0, _possibleConstructorReturn3.default)(this, (BazaarvoiceReview.__proto__ || (0, _getPrototypeOf2.default)(BazaarvoiceReview)).apply(this, arguments));
    }

    (0, _createClass3.default)(BazaarvoiceReview, [{
        key: 'componentDidMount',
        value: function componentDidMount() {
            var productId = this.props.productId;


            var $BV = window.$BV;

            // We can't really test that a function that doesn't exist wasn't called
            /* istanbul ignore else */
            if ($BV) {
                $BV.ui('rr', 'show_reviews', {
                    productId: productId
                });
            }
        }
    }, {
        key: 'render',
        value: function render() {
            var _props = this.props,
                className = _props.className,
                containerId = _props.containerId;


            var classes = (0, _classnames2.default)('pw-bazaarvoice-review', className);

            return _react2.default.createElement(
                'div',
                { className: classes },
                _react2.default.createElement('div', { id: containerId })
            );
        }
    }]);
    return BazaarvoiceReview;
}(_react2.default.Component);

BazaarvoiceReview.defaultProps = {
    containerId: 'BVRRContainer'
};

BazaarvoiceReview.propTypes = {
    /**
     * The src for the Bazaarvoice API script
     */
    apiSrc: _propTypes2.default.string.isRequired,

    /**
     * The product ID the reviews should be shown for
     */
    productId: _propTypes2.default.string.isRequired,

    /**
     * Adds values to the `class` attribute of the root element
     */
    className: _propTypes2.default.string,

    /**
     * This id is added to the element Bazaarvoice should render the
     * review content in
     */
    containerId: _propTypes2.default.string

};

exports.default = (0, _bazaarvoiceWrapper2.default)(BazaarvoiceReview);
