// mobify-integration-manager/dist/integration-manager/api/product-search/types.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.KeyedProductSearch = exports.ProductSearchKey = exports.ProductSearch = exports.ProductSearchHit = undefined;

var _runtypes = require('runtypes');

var Runtypes = _interopRequireWildcard(_runtypes);

var _types = require('../types');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2017 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var ProductSearchHit = exports.ProductSearchHit = Runtypes.Record({
    productId: _types.Text,
    productName: _types.Text,
    link: _types.Text
});

var ProductSearch = exports.ProductSearch = Runtypes.Record({
    products: Runtypes.Array(ProductSearchHit),
    start: Runtypes.Number,
    total: Runtypes.Number,
    query: _types.Text
});

var ProductSearchKey = exports.ProductSearchKey = _types.Text;

var KeyedProductSearch = exports.KeyedProductSearch = Runtypes.Dictionary(ProductSearch, ProductSearchKey);
