// mobify-integration-manager/dist/integration-manager/api/checkout/types.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.Address = exports.ShippingMethods = exports.LocationList = undefined;

var _runtypes = require('runtypes');

var Runtypes = _interopRequireWildcard(_runtypes);

var _types = require('../types');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */
/* Copyright (c) 2017 Mobify Research & Development Inc. All rights reserved. */
/* * *  *  * *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  *  * */

var CountryID = _types.Identifier;

var Country = Runtypes.Record({
    id: CountryID,
    label: _types.Text,
    regionRequired: Runtypes.Boolean,
    postcodeRequired: Runtypes.Boolean
});

var Region = Runtypes.Record({
    countryId: CountryID,
    id: _types.Identifier,
    label: _types.Text
});

var LocationList = exports.LocationList = Runtypes.Record({
    countries: Runtypes.Array(Country),
    regions: Runtypes.Array(Region)
});

var ShippingMethod = Runtypes.Record({
    label: _types.Text,
    cost: _types.Money,
    id: _types.Identifier
});

var ShippingMethods = exports.ShippingMethods = Runtypes.Array(ShippingMethod);

var Address = exports.Address = Runtypes.Record({
    firstname: _types.Text,
    lastname: _types.Text,
    addressLine1: _types.Text,
    city: _types.Text,
    countryId: CountryID,
    postcode: _types.Text,
    telephone: _types.Text
}).And(Runtypes.Partial({
    addressLine2: (0, _types.Nullable)(_types.Text), // eslint-disable-line new-cap
    // We expect one of these two "region" fields to be non-null
    regionId: (0, _types.Nullable)(_types.Identifier), // eslint-disable-line new-cap
    region: (0, _types.Nullable)(_types.Text) // eslint-disable-line new-cap
}));
